// @btc-vision/ecpair/build/capability.js
/**
 * Bitfield flags describing the cryptographic operations an
 * {@link ECPairSigner} can perform.
 *
 * Combine with bitwise OR (`|`) and test with bitwise AND (`&`).
 *
 * @example
 * ```ts
 * const caps = signer.capabilities;
 * if (caps & SignerCapability.SchnorrSign) {
 *     // Schnorr signing is available
 * }
 * ```
 */
export const SignerCapability = {
    /** Can produce ECDSA signatures (requires a private key). */
    EcdsaSign: 1 << 0,
    /** Can verify ECDSA signatures. */
    EcdsaVerify: 1 << 1,
    /** Can produce BIP-340 Schnorr signatures (requires a private key and backend support). */
    SchnorrSign: 1 << 2,
    /** Can verify BIP-340 Schnorr signatures (requires backend support). */
    SchnorrVerify: 1 << 3,
    /** Can export the raw private key bytes. */
    PrivateKeyExport: 1 << 4,
    /** Can derive a tweaked child key via Taproot-style tweaking. */
    PublicKeyTweak: 1 << 5,
    /** Supports BIP-32 hierarchical deterministic derivation. */
    HdDerivation: 1 << 6,
};
