// @btc-vision/ecpair/build/wif.js
import { createPrivateKey } from './types.js';
import { decode, encode } from '@btc-vision/wif';
/**
 * Encodes a private key into Wallet Import Format (WIF / Base58Check).
 *
 * @param privateKey - 32-byte private key to encode.
 * @param compressed - `true` to flag the key as compressed.
 * @param network - Network whose {@link Network.wif | wif} version byte to use.
 * @returns The WIF-encoded string.
 */
export function encodeWIF(privateKey, compressed, network) {
    return encode({
        version: network.wif,
        privateKey,
        compressed,
    });
}
/**
 * Decodes a WIF-encoded private key string.
 *
 * When `network` is a single {@link Network}, the decoded version byte must
 * match exactly or an error is thrown.  When `network` is an array, the first
 * network whose {@link Network.wif | wif} byte matches is used.
 *
 * @param wifString - The WIF string to decode.
 * @param network - One or more candidate networks.
 * @returns The decoded private key, compression flag, and matched network.
 * @throws If no network matches the decoded version byte.
 */
export function decodeWIF(wifString, network) {
    const decoded = decode(wifString);
    const version = decoded.version;
    if (Array.isArray(network)) {
        const nets = network;
        const matched = nets.find((n) => n.wif === version);
        if (!matched)
            throw new Error('Unknown network version');
        return {
            privateKey: createPrivateKey(decoded.privateKey),
            compressed: decoded.compressed,
            network: matched,
        };
    }
    const net = network;
    if (version !== net.wif) {
        throw new Error('Invalid network version');
    }
    return {
        privateKey: createPrivateKey(decoded.privateKey),
        compressed: decoded.compressed,
        network: net,
    };
}
