// @btc-vision/opnet-transform/transform/StrToAbiType.ts
import { ABIDataTypes } from '@btc-vision/transaction';

export const StrToAbiType: { [key: string]: ABIDataTypes } = {
    // ABI-style type names (canonical camelCase — must match opnet's TypeToStr.ts)
    address: ABIDataTypes.ADDRESS,
    extendedAddress: ABIDataTypes.EXTENDED_ADDRESS,
    bool: ABIDataTypes.BOOL,
    bytes: ABIDataTypes.BYTES,
    uint256: ABIDataTypes.UINT256,
    uint128: ABIDataTypes.UINT128,
    uint64: ABIDataTypes.UINT64,
    uint32: ABIDataTypes.UINT32,
    uint16: ABIDataTypes.UINT16,
    uint8: ABIDataTypes.UINT8,
    int128: ABIDataTypes.INT128,
    int64: ABIDataTypes.INT64,
    int32: ABIDataTypes.INT32,
    int16: ABIDataTypes.INT16,
    int8: ABIDataTypes.INT8,
    string: ABIDataTypes.STRING,
    bytes4: ABIDataTypes.BYTES4,
    bytes32: ABIDataTypes.BYTES32,
    schnorrSignature: ABIDataTypes.SCHNORR_SIGNATURE,
    'tuple(address,uint256)[]': ABIDataTypes.ADDRESS_UINT256_TUPLE,
    'tuple(extendedAddress,uint256)[]': ABIDataTypes.EXTENDED_ADDRESS_UINT256_TUPLE,
    'address[]': ABIDataTypes.ARRAY_OF_ADDRESSES,
    'extendedAddress[]': ABIDataTypes.ARRAY_OF_EXTENDED_ADDRESSES,
    'uint256[]': ABIDataTypes.ARRAY_OF_UINT256,
    'uint128[]': ABIDataTypes.ARRAY_OF_UINT128,
    'uint64[]': ABIDataTypes.ARRAY_OF_UINT64,
    'uint32[]': ABIDataTypes.ARRAY_OF_UINT32,
    'uint16[]': ABIDataTypes.ARRAY_OF_UINT16,
    'uint8[]': ABIDataTypes.ARRAY_OF_UINT8,
    'bytes[]': ABIDataTypes.ARRAY_OF_BYTES,
    'buffer[]': ABIDataTypes.ARRAY_OF_BUFFERS,
    'string[]': ABIDataTypes.ARRAY_OF_STRING,

    // AssemblyScript type aliases
    u256: ABIDataTypes.UINT256,
    u128: ABIDataTypes.UINT128,
    u64: ABIDataTypes.UINT64,
    u32: ABIDataTypes.UINT32,
    u16: ABIDataTypes.UINT16,
    u8: ABIDataTypes.UINT8,
    i128: ABIDataTypes.INT128,
    i64: ABIDataTypes.INT64,
    i32: ABIDataTypes.INT32,
    i16: ABIDataTypes.INT16,
    i8: ABIDataTypes.INT8,

    'AddressMap<u256>': ABIDataTypes.ADDRESS_UINT256_TUPLE,
    'ExtendedAddressMap<u256>': ABIDataTypes.EXTENDED_ADDRESS_UINT256_TUPLE,
    Address: ABIDataTypes.ADDRESS,
    ExtendedAddress: ABIDataTypes.EXTENDED_ADDRESS,
    SchnorrSignature: ABIDataTypes.SCHNORR_SIGNATURE,
    'u256[]': ABIDataTypes.ARRAY_OF_UINT256,
    'u128[]': ABIDataTypes.ARRAY_OF_UINT128,
    'u64[]': ABIDataTypes.ARRAY_OF_UINT64,
    'u32[]': ABIDataTypes.ARRAY_OF_UINT32,
    'u16[]': ABIDataTypes.ARRAY_OF_UINT16,
    'u8[]': ABIDataTypes.ARRAY_OF_UINT8,
    'Address[]': ABIDataTypes.ARRAY_OF_ADDRESSES,
    'ExtendedAddress[]': ABIDataTypes.ARRAY_OF_EXTENDED_ADDRESSES,
    'Uint8Array[]': ABIDataTypes.ARRAY_OF_BYTES,
    Uint8Array: ABIDataTypes.BYTES,
    boolean: ABIDataTypes.BOOL,
};

// Explicit canonical reverse mapping — mirrors opnet's TypeToStr.ts exactly.
// This avoids iteration-order fragility from auto-computing the reverse of StrToAbiType.
export const AbiTypeToStr: { [key in ABIDataTypes]: string } = {
    [ABIDataTypes.ADDRESS]: 'address',
    [ABIDataTypes.EXTENDED_ADDRESS]: 'extendedAddress',
    [ABIDataTypes.BOOL]: 'bool',
    [ABIDataTypes.BYTES]: 'bytes',
    [ABIDataTypes.BYTES32]: 'bytes32',
    [ABIDataTypes.BYTES4]: 'bytes4',
    [ABIDataTypes.UINT256]: 'uint256',
    [ABIDataTypes.UINT128]: 'uint128',
    [ABIDataTypes.UINT64]: 'uint64',
    [ABIDataTypes.UINT32]: 'uint32',
    [ABIDataTypes.UINT16]: 'uint16',
    [ABIDataTypes.UINT8]: 'uint8',
    [ABIDataTypes.INT128]: 'int128',
    [ABIDataTypes.INT64]: 'int64',
    [ABIDataTypes.INT32]: 'int32',
    [ABIDataTypes.INT16]: 'int16',
    [ABIDataTypes.INT8]: 'int8',
    [ABIDataTypes.STRING]: 'string',
    [ABIDataTypes.ADDRESS_UINT256_TUPLE]: 'tuple(address,uint256)[]',
    [ABIDataTypes.EXTENDED_ADDRESS_UINT256_TUPLE]: 'tuple(extendedAddress,uint256)[]',
    [ABIDataTypes.SCHNORR_SIGNATURE]: 'schnorrSignature',
    [ABIDataTypes.ARRAY_OF_ADDRESSES]: 'address[]',
    [ABIDataTypes.ARRAY_OF_EXTENDED_ADDRESSES]: 'extendedAddress[]',
    [ABIDataTypes.ARRAY_OF_UINT256]: 'uint256[]',
    [ABIDataTypes.ARRAY_OF_UINT128]: 'uint128[]',
    [ABIDataTypes.ARRAY_OF_UINT64]: 'uint64[]',
    [ABIDataTypes.ARRAY_OF_UINT32]: 'uint32[]',
    [ABIDataTypes.ARRAY_OF_UINT16]: 'uint16[]',
    [ABIDataTypes.ARRAY_OF_UINT8]: 'uint8[]',
    [ABIDataTypes.ARRAY_OF_BYTES]: 'bytes[]',
    [ABIDataTypes.ARRAY_OF_STRING]: 'string[]',
    [ABIDataTypes.ARRAY_OF_BUFFERS]: 'buffer[]',
};
