// @btc-vision/bitcoin/build/workers/psbt-parallel.js
/**
 * PSBT parallel signing integration.
 *
 * Provides helper functions to sign PSBT inputs in parallel using worker threads.
 *
 * @example
 * ```typescript
 * import { Psbt } from '@btc-vision/bitcoin';
 * import { signPsbtParallel, WorkerSigningPool } from '@btc-vision/bitcoin/workers';
 *
 * // Initialize pool once at app startup
 * const pool = WorkerSigningPool.getInstance();
 * pool.preserveWorkers();
 *
 * // Create and populate PSBT
 * const psbt = new Psbt();
 * psbt.addInput(...);
 * psbt.addOutput(...);
 *
 * // Sign all inputs in parallel
 * await signPsbtParallel(psbt, keyPair, pool);
 *
 * // Finalize and extract
 * psbt.finalizeAllInputs();
 * const tx = psbt.extractTransaction();
 * ```
 *
 * @packageDocumentation
 */
import { Transaction } from '../transaction.js';
import { SignatureType } from './types.js';
import { toXOnly } from '../pubkey.js';
import { isTaprootInput, serializeTaprootSignature } from '../psbt/bip371.js';
import * as bscript from '../script.js';
/**
 * Signs all PSBT inputs in parallel using worker threads.
 *
 * This is the main entry point for parallel PSBT signing.
 * All inputs that can be signed with the provided key pair
 * will be signed simultaneously.
 *
 * SECURITY: Private keys are isolated per-worker and zeroed after signing.
 *
 * @param psbt - The PSBT to sign
 * @param keyPair - Key pair with getPrivateKey() method
 * @param poolOrConfig - Existing pool instance or configuration for new pool
 * @param options - Signing options
 * @returns Promise resolving to signing result
 *
 * @example
 * ```typescript
 * // With existing pool (recommended for multiple operations)
 * const pool = WorkerSigningPool.getInstance();
 * pool.preserveWorkers();
 *
 * const result = await signPsbtParallel(psbt, keyPair, pool);
 *
 * if (result.success) {
 *     psbt.finalizeAllInputs();
 * }
 *
 * // With inline pool (creates and destroys pool)
 * const result = await signPsbtParallel(psbt, keyPair, { workerCount: 4 });
 * ```
 */
export async function signPsbtParallel(psbt, keyPair, poolOrConfig, options = {}) {
    // Get or create pool
    let pool;
    let shouldShutdown = false;
    if (poolOrConfig && 'signBatch' in poolOrConfig) {
        pool = poolOrConfig;
    }
    else {
        const { WorkerSigningPool } = await import('./WorkerSigningPool.js');
        pool = WorkerSigningPool.getInstance(poolOrConfig);
        if (!pool.isPreservingWorkers) {
            shouldShutdown = true;
        }
    }
    try {
        // Initialize pool if needed
        await pool.initialize();
        // Prepare signing tasks
        const tasks = prepareSigningTasks(psbt, keyPair, options);
        if (tasks.length === 0) {
            return {
                success: true,
                signatures: new Map(),
                errors: new Map(),
                durationMs: 0,
            };
        }
        // Sign in parallel
        const result = await pool.signBatch(tasks, keyPair);
        // Apply signatures to PSBT
        if (result.success) {
            applySignaturesToPsbt(psbt, result, keyPair);
        }
        return result;
    }
    finally {
        if (shouldShutdown) {
            await pool.shutdown();
        }
    }
}
/**
 * Prepares signing tasks from a PSBT.
 *
 * Analyzes each input and creates a signing task for inputs
 * that can be signed with the provided key pair.
 *
 * @param psbt - The PSBT to analyze
 * @param keyPair - Key pair to check against
 * @param options - Signing options
 * @returns Array of signing tasks
 */
export function prepareSigningTasks(psbt, keyPair, options = {}) {
    const tasks = [];
    const inputs = psbt.data.inputs;
    const pubkey = keyPair.publicKey;
    for (let i = 0; i < inputs.length; i++) {
        const input = inputs[i];
        // Check if this input can be signed with this key
        if (!psbt.inputHasPubkey(i, pubkey)) {
            continue;
        }
        // Determine if Taproot
        if (isTaprootInput(input)) {
            // Get Taproot hashes for signing
            const taprootTasks = prepareTaprootTasks(psbt, i, input, keyPair, options);
            tasks.push(...taprootTasks);
        }
        else {
            // Legacy/SegWit signing
            const legacyTask = prepareLegacyTask(psbt, i, input, keyPair, options);
            if (legacyTask) {
                tasks.push(legacyTask);
            }
        }
    }
    return tasks;
}
/**
 * Prepares signing tasks for a Taproot input.
 */
function prepareTaprootTasks(psbt, inputIndex, input, keyPair, options) {
    const tasks = [];
    try {
        const hashesForSig = psbt.checkTaprootHashesForSig(inputIndex, input, keyPair, options.tapLeafHash, options.sighashTypes);
        for (const { hash, leafHash } of hashesForSig) {
            tasks.push({
                taskId: `taproot-${inputIndex}-${leafHash ? 'script' : 'key'}`,
                inputIndex,
                hash,
                signatureType: SignatureType.Schnorr,
                sighashType: input.sighashType ?? Transaction.SIGHASH_DEFAULT,
                leafHash,
            });
        }
    }
    catch {
        // Input cannot be signed with this key
    }
    return tasks;
}
/**
 * Prepares a signing task for a legacy/SegWit input.
 */
function prepareLegacyTask(_psbt, _inputIndex, input, _keyPair, options) {
    try {
        // Get hash for signing - we need to access the internal method
        // This is a simplified version; full implementation would need cache access
        const sighashType = input.sighashType ?? Transaction.SIGHASH_ALL;
        const allowedTypes = options.sighashTypes ?? [Transaction.SIGHASH_ALL];
        if (!allowedTypes.includes(sighashType)) {
            return null;
        }
        // Note: In production, we'd need to compute the hash properly
        // This requires access to PSBT internals (cache, etc.)
        // For now, we return a placeholder - the actual implementation
        // would integrate with psbt._signInput internals
        return null; // Placeholder - full implementation needs PSBT internals
    }
    catch {
        return null;
    }
}
/**
 * Applies parallel signing results to a PSBT.
 *
 * @param psbt - The PSBT to update
 * @param result - Signing results from parallel signing
 * @param keyPair - Key pair used for signing
 */
export function applySignaturesToPsbt(psbt, result, keyPair) {
    const pubkey = keyPair.publicKey;
    for (const [inputIndex, sigResult] of result.signatures) {
        const input = psbt.data.inputs[inputIndex];
        if (sigResult.signatureType === SignatureType.Schnorr) {
            // Taproot signature
            if (sigResult.leafHash) {
                // Script-path signature
                const tapScriptSig = [
                    {
                        pubkey: toXOnly(pubkey),
                        signature: serializeTaprootSignature(sigResult.signature, input.sighashType),
                        leafHash: sigResult.leafHash,
                    },
                ];
                psbt.data.updateInput(inputIndex, { tapScriptSig: tapScriptSig });
            }
            else {
                // Key-path signature
                const tapKeySig = serializeTaprootSignature(sigResult.signature, input.sighashType);
                psbt.data.updateInput(inputIndex, { tapKeySig: tapKeySig });
            }
        }
        else {
            // ECDSA signature
            const encodedSig = bscript.signature.encode(sigResult.signature, input.sighashType ?? Transaction.SIGHASH_ALL);
            const partialSig = [
                {
                    pubkey: Uint8Array.from(pubkey),
                    signature: encodedSig,
                },
            ];
            psbt.data.updateInput(inputIndex, { partialSig });
        }
    }
}
// Transaction is already imported above for sighash constants
