// @btc-vision/bitcoin/build/workers/types.js
/**
 * Worker thread types for parallel signing operations.
 *
 * SECURITY NOTES:
 * - Private keys are NEVER shared via SharedArrayBuffer
 * - Keys are cloned via postMessage (isolated per worker)
 * - Keys are zeroed immediately after signing
 * - Workers are pooled for performance but keys don't persist
 *
 * @packageDocumentation
 */
/**
 * Signature type enum for worker operations.
 */
export const SignatureType = {
    /** ECDSA signature (secp256k1) */
    ECDSA: 0,
    /** Schnorr signature (BIP340) */
    Schnorr: 1,
};
/**
 * Type guard for error responses.
 */
export function isSigningError(response) {
    return response.type === 'error';
}
/**
 * Type guard for success responses.
 */
export function isSigningResult(response) {
    return response.type === 'result';
}
/**
 * Type guard for batch result responses.
 */
export function isBatchResult(response) {
    return response.type === 'batchResult';
}
/**
 * Type guard for ready responses.
 */
export function isWorkerReady(response) {
    return response.type === 'ready';
}
/**
 * Worker state for pool management.
 */
export const WorkerState = {
    /** Worker is initializing */
    Initializing: 0,
    /** Worker is ready and idle */
    Idle: 1,
    /** Worker is processing a task */
    Busy: 2,
    /** Worker is shutting down */
    ShuttingDown: 3,
    /** Worker has terminated */
    Terminated: 4,
};
