// @btc-vision/bitcoin/scripts/bundle-ecc.ts
/**
 * Build script to bundle @noble/secp256k1 + hashes for embedding in workers.
 *
 * This creates a self-contained IIFE that can be executed in a worker
 * without any network requests or module system dependencies.
 *
 * Usage: npx tsx scripts/bundle-ecc.ts
 */

import { build } from 'esbuild';
import { writeFileSync, mkdirSync } from 'fs';
import { dirname, resolve } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const projectRoot = resolve(__dirname, '..');

async function bundleEcc(): Promise<void> {
    console.log('Bundling @noble/secp256k1 + hashes for worker embedding...');

    // Create a wrapper that exports everything we need
    const entryCode = `
        import * as secp from '@noble/secp256k1';
        import { sha256 } from '@noble/hashes/sha2.js';
        import { hmac } from '@noble/hashes/hmac.js';

        // Configure hashes for sync operations
        secp.hashes.sha256 = sha256;
        secp.hashes.hmacSha256 = (key, ...msgs) => {
            const merged = new Uint8Array(msgs.reduce((acc, m) => acc + m.length, 0));
            let offset = 0;
            for (const m of msgs) {
                merged.set(m, offset);
                offset += m.length;
            }
            return hmac(sha256, key, merged);
        };

        export { secp, sha256, hmac };
    `;

    // Write temp entry file
    const tempEntry = resolve(projectRoot, 'scripts/.temp-ecc-entry.js');
    writeFileSync(tempEntry, entryCode);

    // Bundle everything as IIFE
    const result = await build({
        entryPoints: [tempEntry],
        bundle: true,
        format: 'iife',
        globalName: 'nobleBundle',
        platform: 'browser',
        target: 'es2020',
        minify: true,
        write: false,
    });

    // Clean up temp file
    const { unlinkSync } = await import('fs');
    unlinkSync(tempEntry);

    if (result.outputFiles.length === 0) {
        throw new Error('esbuild produced no output');
    }

    const bundledCode = result.outputFiles[0].text;

    // Generate TypeScript file with the bundled code as a constant
    const outputContent = `/**
 * Bundled @noble/secp256k1 + hashes for worker embedding.
 *
 * AUTO-GENERATED FILE - DO NOT EDIT
 * Generated by: scripts/bundle-ecc.ts
 *
 * This embeds the entire @noble/secp256k1 library with sha256/hmac
 * as an IIFE string that can be executed in a Web Worker without network requests.
 *
 * @packageDocumentation
 */

/**
 * Bundled @noble/secp256k1 + hashes library as an IIFE string.
 *
 * When executed, this creates a global \`nobleBundle\` object with:
 * - secp: The full @noble/secp256k1 module (with hashes configured)
 * - sha256: The sha256 hash function
 * - hmac: The hmac function
 */
export const ECC_BUNDLE = ${JSON.stringify(bundledCode)};

/**
 * Size of the bundled code in bytes.
 */
export const ECC_BUNDLE_SIZE = ${bundledCode.length};
`;

    // Ensure output directory exists
    const outputPath = resolve(projectRoot, 'src/workers/ecc-bundle.ts');
    mkdirSync(dirname(outputPath), { recursive: true });

    // Write the output file
    writeFileSync(outputPath, outputContent, 'utf-8');

    console.log('Bundle created: ' + outputPath);
    console.log('Bundle size: ' + (bundledCode.length / 1024).toFixed(2) + ' KB');
}

bundleEcc().catch((error) => {
    console.error('Failed to bundle ECC library:', error);
    process.exit(1);
});
