// @btc-vision/transaction/build/abi/ABIDataTypes.js
export var ABIDataTypes;
(function (ABIDataTypes) {
    // Unsigned integers
    ABIDataTypes["UINT8"] = "UINT8";
    ABIDataTypes["UINT16"] = "UINT16";
    ABIDataTypes["UINT32"] = "UINT32";
    ABIDataTypes["UINT64"] = "UINT64";
    ABIDataTypes["UINT128"] = "UINT128";
    ABIDataTypes["UINT256"] = "UINT256";
    // Signed integers
    ABIDataTypes["INT8"] = "INT8";
    ABIDataTypes["INT16"] = "INT16";
    ABIDataTypes["INT32"] = "INT32";
    ABIDataTypes["INT64"] = "INT64";
    ABIDataTypes["INT128"] = "INT128";
    // Basic types
    ABIDataTypes["BOOL"] = "BOOL";
    ABIDataTypes["ADDRESS"] = "ADDRESS";
    ABIDataTypes["EXTENDED_ADDRESS"] = "EXTENDED_ADDRESS";
    ABIDataTypes["STRING"] = "STRING";
    ABIDataTypes["BYTES4"] = "BYTES4";
    ABIDataTypes["BYTES32"] = "BYTES32";
    ABIDataTypes["BYTES"] = "BYTES";
    // Tuples/Maps
    ABIDataTypes["ADDRESS_UINT256_TUPLE"] = "ADDRESS_UINT256_TUPLE";
    ABIDataTypes["EXTENDED_ADDRESS_UINT256_TUPLE"] = "EXTENDED_ADDRESS_UINT256_TUPLE";
    // Signatures
    ABIDataTypes["SCHNORR_SIGNATURE"] = "SCHNORR_SIGNATURE";
    // Arrays
    ABIDataTypes["ARRAY_OF_ADDRESSES"] = "ARRAY_OF_ADDRESSES";
    ABIDataTypes["ARRAY_OF_EXTENDED_ADDRESSES"] = "ARRAY_OF_EXTENDED_ADDRESSES";
    ABIDataTypes["ARRAY_OF_UINT256"] = "ARRAY_OF_UINT256";
    ABIDataTypes["ARRAY_OF_UINT128"] = "ARRAY_OF_UINT128";
    ABIDataTypes["ARRAY_OF_UINT64"] = "ARRAY_OF_UINT64";
    ABIDataTypes["ARRAY_OF_UINT32"] = "ARRAY_OF_UINT32";
    ABIDataTypes["ARRAY_OF_UINT16"] = "ARRAY_OF_UINT16";
    ABIDataTypes["ARRAY_OF_UINT8"] = "ARRAY_OF_UINT8";
    ABIDataTypes["ARRAY_OF_STRING"] = "ARRAY_OF_STRING";
    ABIDataTypes["ARRAY_OF_BYTES"] = "ARRAY_OF_BYTES";
    ABIDataTypes["ARRAY_OF_BUFFERS"] = "ARRAY_OF_BUFFERS";
})(ABIDataTypes || (ABIDataTypes = {}));
