// @btc-vision/transaction/build/abi/AbiTypes.js
import { ABIDataTypes } from './ABIDataTypes.js';
/**
 * Canonical string → ABIDataTypes mapping.
 * Only includes canonical ABI-style names (not AssemblyScript aliases).
 */
export const StrToAbiType = {
    address: ABIDataTypes.ADDRESS,
    extendedAddress: ABIDataTypes.EXTENDED_ADDRESS,
    bool: ABIDataTypes.BOOL,
    bytes: ABIDataTypes.BYTES,
    uint256: ABIDataTypes.UINT256,
    uint128: ABIDataTypes.UINT128,
    uint64: ABIDataTypes.UINT64,
    uint32: ABIDataTypes.UINT32,
    uint16: ABIDataTypes.UINT16,
    uint8: ABIDataTypes.UINT8,
    int128: ABIDataTypes.INT128,
    int64: ABIDataTypes.INT64,
    int32: ABIDataTypes.INT32,
    int16: ABIDataTypes.INT16,
    int8: ABIDataTypes.INT8,
    string: ABIDataTypes.STRING,
    bytes4: ABIDataTypes.BYTES4,
    bytes32: ABIDataTypes.BYTES32,
    schnorrSignature: ABIDataTypes.SCHNORR_SIGNATURE,
    'tuple(address,uint256)[]': ABIDataTypes.ADDRESS_UINT256_TUPLE,
    'tuple(extendedAddress,uint256)[]': ABIDataTypes.EXTENDED_ADDRESS_UINT256_TUPLE,
    'address[]': ABIDataTypes.ARRAY_OF_ADDRESSES,
    'extendedAddress[]': ABIDataTypes.ARRAY_OF_EXTENDED_ADDRESSES,
    'uint256[]': ABIDataTypes.ARRAY_OF_UINT256,
    'uint128[]': ABIDataTypes.ARRAY_OF_UINT128,
    'uint64[]': ABIDataTypes.ARRAY_OF_UINT64,
    'uint32[]': ABIDataTypes.ARRAY_OF_UINT32,
    'uint16[]': ABIDataTypes.ARRAY_OF_UINT16,
    'uint8[]': ABIDataTypes.ARRAY_OF_UINT8,
    'bytes[]': ABIDataTypes.ARRAY_OF_BYTES,
    'buffer[]': ABIDataTypes.ARRAY_OF_BUFFERS,
    'string[]': ABIDataTypes.ARRAY_OF_STRING,
    boolean: ABIDataTypes.BOOL,
};
/**
 * Canonical reverse mapping: ABIDataTypes → canonical string.
 */
export const AbiTypeToStr = {
    [ABIDataTypes.ADDRESS]: 'address',
    [ABIDataTypes.EXTENDED_ADDRESS]: 'extendedAddress',
    [ABIDataTypes.BOOL]: 'bool',
    [ABIDataTypes.BYTES]: 'bytes',
    [ABIDataTypes.BYTES32]: 'bytes32',
    [ABIDataTypes.BYTES4]: 'bytes4',
    [ABIDataTypes.UINT256]: 'uint256',
    [ABIDataTypes.UINT128]: 'uint128',
    [ABIDataTypes.UINT64]: 'uint64',
    [ABIDataTypes.UINT32]: 'uint32',
    [ABIDataTypes.UINT16]: 'uint16',
    [ABIDataTypes.UINT8]: 'uint8',
    [ABIDataTypes.INT128]: 'int128',
    [ABIDataTypes.INT64]: 'int64',
    [ABIDataTypes.INT32]: 'int32',
    [ABIDataTypes.INT16]: 'int16',
    [ABIDataTypes.INT8]: 'int8',
    [ABIDataTypes.STRING]: 'string',
    [ABIDataTypes.ADDRESS_UINT256_TUPLE]: 'tuple(address,uint256)[]',
    [ABIDataTypes.EXTENDED_ADDRESS_UINT256_TUPLE]: 'tuple(extendedAddress,uint256)[]',
    [ABIDataTypes.SCHNORR_SIGNATURE]: 'schnorrSignature',
    [ABIDataTypes.ARRAY_OF_ADDRESSES]: 'address[]',
    [ABIDataTypes.ARRAY_OF_EXTENDED_ADDRESSES]: 'extendedAddress[]',
    [ABIDataTypes.ARRAY_OF_UINT256]: 'uint256[]',
    [ABIDataTypes.ARRAY_OF_UINT128]: 'uint128[]',
    [ABIDataTypes.ARRAY_OF_UINT64]: 'uint64[]',
    [ABIDataTypes.ARRAY_OF_UINT32]: 'uint32[]',
    [ABIDataTypes.ARRAY_OF_UINT16]: 'uint16[]',
    [ABIDataTypes.ARRAY_OF_UINT8]: 'uint8[]',
    [ABIDataTypes.ARRAY_OF_BYTES]: 'bytes[]',
    [ABIDataTypes.ARRAY_OF_STRING]: 'string[]',
    [ABIDataTypes.ARRAY_OF_BUFFERS]: 'buffer[]',
};
