// @btc-vision/bitcoin/build/index.js
import './env.js';
import * as networks from './networks.js';
import * as address from './address.js';
import * as payments from './payments/index.js';
import * as script from './script.js';
import * as crypto from './crypto.js';
import * as Transaction from './transaction.js';
// Custom error types
import { AddressError, BitcoinError, EccError, InvalidInputError, InvalidOutputError, PsbtError, ScriptError, SignatureError, ValidationError, } from './errors.js';
export * as address from './address.js';
export * as crypto from './crypto.js';
export * as networks from './networks.js';
export * as payments from './payments/index.js';
export * as script from './script.js';
export { Block } from './block.js';
/** @hidden */
export * from './crypto.js';
export { Psbt, PsbtCache, PsbtSigner, PsbtFinalizer, PsbtTransaction, transactionFromBuffer, getFinalScripts, prepareFinalScripts, } from './psbt.js';
/** @hidden */
export { opcodes } from './opcodes.js';
export { Transaction } from './transaction.js';
/** @hidden */
export { initEccLib, getEccLib, EccContext } from './ecc/context.js';
export { PaymentType } from './payments/index.js';
export * from './psbt/bip371.js';
export * from './address.js';
export { toHex, fromHex, isHex, fromBase64, toBase64, concat, equals, compare, isZero, clone, reverse, reverseCopy, alloc, xor, fromUtf8, toUtf8, varuint, } from './io/index.js';
export * from './payments/bip341.js';
export * from './psbt/psbtutils.js';
export { toXOnly, decompressPublicKey, pubkeysMatch, } from './pubkey.js';
export { TAPLEAF_VERSION_MASK } from './types.js';
export { isPrivateKey, isSchnorrSignature, isSignature, isScript, isBytes32, isBytes20, isPoint, isXOnlyPublicKey, isSatoshi, assertXOnlyPublicKey, assertPrivateKey, toBytes32, toBytes20, toSatoshi, } from './types.js';
export { BitcoinError, ValidationError, InvalidInputError, InvalidOutputError, ScriptError, PsbtError, EccError, AddressError, SignatureError, };
/** All error classes grouped as a namespace */
export const errors = {
    BitcoinError,
    ValidationError,
    InvalidInputError,
    InvalidOutputError,
    ScriptError,
    PsbtError,
    EccError,
    AddressError,
    SignatureError,
};
// Worker-based parallel signing (browser-safe shared exports only)
// For createSigningPool / detectRuntime, use '@btc-vision/bitcoin/workers' subpath export
// which routes to the correct platform-specific implementation.
export { WorkerSigningPool, getSigningPool, SignatureType, signPsbtParallel, prepareSigningTasks, applySignaturesToPsbt, } from './workers/index.shared.js';
const bitcoin = {
    networks,
    address,
    payments,
    script,
    crypto,
    Transaction,
};
export default bitcoin;
