// @btc-vision/transaction/build/transaction/browser/WalletNetworks.js
export var WalletNetworks;
(function (WalletNetworks) {
    WalletNetworks["Mainnet"] = "mainnet";
    WalletNetworks["Testnet"] = "testnet";
    WalletNetworks["OpnetTestnet"] = "opnetTestnet";
    WalletNetworks["Regtest"] = "regtest";
})(WalletNetworks || (WalletNetworks = {}));
export var UnisatChainType;
(function (UnisatChainType) {
    UnisatChainType["BITCOIN_MAINNET"] = "BITCOIN_MAINNET";
    UnisatChainType["BITCOIN_TESTNET"] = "BITCOIN_TESTNET";
    UnisatChainType["BITCOIN_TESTNET4"] = "BITCOIN_TESTNET4";
    UnisatChainType["BITCOIN_REGTEST"] = "BITCOIN_REGTEST";
    UnisatChainType["BITCOIN_SIGNET"] = "BITCOIN_SIGNET";
    UnisatChainType["OPNET_TESTNET"] = "OPNET_TESTNET";
    UnisatChainType["FRACTAL_BITCOIN_MAINNET"] = "FRACTAL_BITCOIN_MAINNET";
    UnisatChainType["FRACTAL_BITCOIN_TESTNET"] = "FRACTAL_BITCOIN_TESTNET";
    UnisatChainType["DOGECOIN_MAINNET"] = "DOGECOIN_MAINNET";
    UnisatChainType["DOGECOIN_TESTNET"] = "DOGECOIN_TESTNET";
    UnisatChainType["DOGECOIN_REGTEST"] = "DOGECOIN_REGTEST";
    UnisatChainType["LITECOIN_MAINNET"] = "LITECOIN_MAINNET";
    UnisatChainType["LITECOIN_TESTNET"] = "LITECOIN_TESTNET";
    UnisatChainType["LITECOIN_REGTEST"] = "LITECOIN_REGTEST";
    UnisatChainType["BITCOINCASH_MAINNET"] = "BITCOINCASH_MAINNET";
    UnisatChainType["BITCOINCASH_TESTNET"] = "BITCOINCASH_TESTNET";
    UnisatChainType["BITCOINCASH_REGTEST"] = "BITCOINCASH_REGTEST";
    UnisatChainType["DASH_MAINNET"] = "DASH_MAINNET";
    UnisatChainType["DASH_TESTNET"] = "DASH_TESTNET";
    UnisatChainType["DASH_REGTEST"] = "DASH_REGTEST";
})(UnisatChainType || (UnisatChainType = {}));
