// opnet/browser/protobuf.js
import { a as getDefaultExportFromCjs } from './vendors.js';

var protobuf_min = {exports: {}};

/*!
 * protobuf.js v8.0.0 (c) 2016, daniel wirtz
 * compiled tue, 16 dec 2025 22:00:06 utc
 * licensed under the bsd-3-clause license
 * see: https://github.com/dcodeio/protobuf.js for details
 */

var hasRequiredProtobuf_min;

function requireProtobuf_min () {
	if (hasRequiredProtobuf_min) return protobuf_min.exports;
	hasRequiredProtobuf_min = 1;
	(function (module) {
		!(function(it) {
		  !(function(r2, e, t2) {
		    var i2 = (function t3(i3) {
		      var n2 = e[i3];
		      return n2 || r2[i3][0].call(n2 = e[i3] = { exports: {} }, t3, n2, n2.exports), n2.exports;
		    })(t2[0]);
		    i2.util.global.protobuf = i2, module && module.exports && (module.exports = i2);
		  })({ 1: [function(t2, i2, n2) {
		    i2.exports = function(t3, i3) {
		      var n3 = Array(arguments.length - 1), s = 0, r2 = 2, o = true;
		      for (; r2 < arguments.length; ) n3[s++] = arguments[r2++];
		      return new Promise(function(r3, e) {
		        n3[s] = function(t4) {
		          if (o) if (o = false, t4) e(t4);
		          else {
		            for (var i4 = Array(arguments.length - 1), n4 = 0; n4 < i4.length; ) i4[n4++] = arguments[n4];
		            r3.apply(null, i4);
		          }
		        };
		        try {
		          t3.apply(i3 || null, n3);
		        } catch (t4) {
		          o && (o = false, e(t4));
		        }
		      });
		    };
		  }, {}], 2: [function(t2, i2, n2) {
		    n2.length = function(t3) {
		      var i3 = t3.length;
		      if (!i3) return 0;
		      for (var n3 = 0; 1 < --i3 % 4 && "=" == (t3[0 | i3] || ""); ) ++n3;
		      return Math.ceil(3 * t3.length) / 4 - n3;
		    };
		    for (var h = Array(64), f = Array(123), r2 = 0; r2 < 64; ) f[h[r2] = r2 < 26 ? r2 + 65 : r2 < 52 ? r2 + 71 : r2 < 62 ? r2 - 4 : r2 - 59 | 43] = r2++;
		    n2.encode = function(t3, i3, n3) {
		      for (var r3, e = null, s = [], o = 0, u = 0; i3 < n3; ) {
		        var f2 = t3[i3++];
		        switch (u) {
		          case 0:
		            s[o++] = h[f2 >> 2], r3 = (3 & f2) << 4, u = 1;
		            break;
		          case 1:
		            s[o++] = h[r3 | f2 >> 4], r3 = (15 & f2) << 2, u = 2;
		            break;
		          case 2:
		            s[o++] = h[r3 | f2 >> 6], s[o++] = h[63 & f2], u = 0;
		        }
		        8191 < o && ((e = e || []).push(String.fromCharCode.apply(String, s)), o = 0);
		      }
		      return u && (s[o++] = h[r3], s[o++] = 61, 1 === u && (s[o++] = 61)), e ? (o && e.push(String.fromCharCode.apply(String, s.slice(0, o))), e.join("")) : String.fromCharCode.apply(String, s.slice(0, o));
		    };
		    var a = "invalid encoding";
		    n2.decode = function(t3, i3, n3) {
		      for (var r3, e = n3, s = 0, o = 0; o < t3.length; ) {
		        var u = t3.charCodeAt(o++);
		        if (61 == u && 1 < s) break;
		        if ((u = f[u]) === it) throw Error(a);
		        switch (s) {
		          case 0:
		            r3 = u, s = 1;
		            break;
		          case 1:
		            i3[n3++] = r3 << 2 | (48 & u) >> 4, r3 = u, s = 2;
		            break;
		          case 2:
		            i3[n3++] = (15 & r3) << 4 | (60 & u) >> 2, r3 = u, s = 3;
		            break;
		          case 3:
		            i3[n3++] = (3 & r3) << 6 | u, s = 0;
		        }
		      }
		      if (1 === s) throw Error(a);
		      return n3 - e;
		    }, n2.test = function(t3) {
		      return /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(t3);
		    };
		  }, {}], 3: [function(t2, i2, n2) {
		    function c(i3, n3) {
		      "string" == typeof i3 && (n3 = i3, i3 = it);
		      var f = [];
		      function h(t3) {
		        if ("string" != typeof t3) {
		          var i4 = a();
		          if (c.verbose && console.log("codegen: " + i4), i4 = "return " + i4, t3) {
		            for (var n4 = Object.keys(t3), r2 = Array(n4.length + 1), e = Array(n4.length), s = 0; s < n4.length; ) r2[s] = n4[s], e[s] = t3[n4[s++]];
		            return r2[s] = i4, Function.apply(null, r2).apply(null, e);
		          }
		          return Function(i4)();
		        }
		        for (var o = Array(arguments.length - 1), u = 0; u < o.length; ) o[u] = arguments[++u];
		        if (u = 0, t3 = t3.replace(/%([%dfijs])/g, function(t4, i5) {
		          var n5 = o[u++];
		          switch (i5) {
		            case "d":
		            case "f":
		              return "" + +("" + n5);
		            case "i":
		              return "" + Math.floor(n5);
		            case "j":
		              return JSON.stringify(n5);
		            case "s":
		              return "" + n5;
		          }
		          return "%";
		        }), u !== o.length) throw Error("parameter count mismatch");
		        return f.push(t3), h;
		      }
		      function a(t3) {
		        return "function " + (t3 || n3 || "") + "(" + (i3 && i3.join(",") || "") + "){\n  " + f.join("\n  ") + "\n}";
		      }
		      return h.toString = a, h;
		    }
		    (i2.exports = c).verbose = false;
		  }, {}], 4: [function(t2, i2, n2) {
		    function r2() {
		      this.i = {};
		    }
		    (i2.exports = r2).prototype.on = function(t3, i3, n3) {
		      return (this.i[t3] || (this.i[t3] = [])).push({ fn: i3, ctx: n3 || this }), this;
		    }, r2.prototype.off = function(t3, i3) {
		      if (t3 === it) this.i = {};
		      else if (i3 === it) this.i[t3] = [];
		      else for (var n3 = this.i[t3], r3 = 0; r3 < n3.length; ) n3[r3].fn === i3 ? n3.splice(r3, 1) : ++r3;
		      return this;
		    }, r2.prototype.emit = function(t3) {
		      var i3 = this.i[t3];
		      if (i3) {
		        for (var n3 = [], r3 = 1; r3 < arguments.length; ) n3.push(arguments[r3++]);
		        for (r3 = 0; r3 < i3.length; ) i3[r3].fn.apply(i3[r3++].ctx, n3);
		      }
		      return this;
		    };
		  }, {}], 5: [function(t2, i2, n2) {
		    i2.exports = u;
		    var s = t2(1), o = t2(7)("fs");
		    function u(n3, r2, e) {
		      return r2 = "function" == typeof r2 ? (e = r2, {}) : r2 || {}, e ? !r2.xhr && o && o.readFile ? o.readFile(n3, function(t3, i3) {
		        return t3 && "undefined" != typeof XMLHttpRequest ? u.xhr(n3, r2, e) : t3 ? e(t3) : e(null, r2.binary ? i3 : i3.toString("utf8"));
		      }) : u.xhr(n3, r2, e) : s(u, this, n3, r2);
		    }
		    u.xhr = function(t3, n3, r2) {
		      var e = new XMLHttpRequest();
		      e.onreadystatechange = function() {
		        if (4 !== e.readyState) return it;
		        if (0 !== e.status && 200 !== e.status) return r2(Error("status " + e.status));
		        if (n3.binary) {
		          if (!(t4 = e.response)) for (var t4 = [], i3 = 0; i3 < e.responseText.length; ++i3) t4.push(255 & e.responseText.charCodeAt(i3));
		          return r2(null, "undefined" != typeof Uint8Array ? new Uint8Array(t4) : t4);
		        }
		        return r2(null, e.responseText);
		      }, n3.binary && ("overrideMimeType" in e && e.overrideMimeType("text/plain; charset=x-user-defined"), e.responseType = "arraybuffer"), e.open("GET", t3), e.send();
		    };
		  }, { 1: 1, 7: 7 }], 6: [function(t2, i2, n2) {
		    function r2(t3) {
		      function i3(t4, i4, n4, r4) {
		        var e2 = i4 < 0 ? 1 : 0;
		        t4(0 === (i4 = e2 ? -i4 : i4) ? 0 < 1 / i4 ? 0 : 2147483648 : isNaN(i4) ? 2143289344 : 34028234663852886e22 < i4 ? (e2 << 31 | 2139095040) >>> 0 : i4 < 11754943508222875e-54 ? (e2 << 31 | Math.round(i4 / 1401298464324817e-60)) >>> 0 : (e2 << 31 | 127 + (t4 = Math.floor(Math.log(i4) / Math.LN2)) << 23 | 8388607 & Math.round(i4 * Math.pow(2, -t4) * 8388608)) >>> 0, n4, r4);
		      }
		      function n3(t4, i4, n4) {
		        t4 = t4(i4, n4), i4 = 2 * (t4 >> 31) + 1, n4 = t4 >>> 23 & 255, t4 &= 8388607;
		        return 255 == n4 ? t4 ? NaN : 1 / 0 * i4 : 0 == n4 ? 1401298464324817e-60 * i4 * t4 : i4 * Math.pow(2, n4 - 150) * (8388608 + t4);
		      }
		      function r3(t4, i4, n4) {
		        u[0] = t4, i4[n4] = f[0], i4[n4 + 1] = f[1], i4[n4 + 2] = f[2], i4[n4 + 3] = f[3];
		      }
		      function e(t4, i4, n4) {
		        u[0] = t4, i4[n4] = f[3], i4[n4 + 1] = f[2], i4[n4 + 2] = f[1], i4[n4 + 3] = f[0];
		      }
		      function s(t4, i4) {
		        return f[0] = t4[i4], f[1] = t4[i4 + 1], f[2] = t4[i4 + 2], f[3] = t4[i4 + 3], u[0];
		      }
		      function o(t4, i4) {
		        return f[3] = t4[i4], f[2] = t4[i4 + 1], f[1] = t4[i4 + 2], f[0] = t4[i4 + 3], u[0];
		      }
		      var u, f, h, a, c;
		      function l(t4, i4, n4, r4, e2, s2) {
		        var o2, u2 = r4 < 0 ? 1 : 0;
		        0 === (r4 = u2 ? -r4 : r4) ? (t4(0, e2, s2 + i4), t4(0 < 1 / r4 ? 0 : 2147483648, e2, s2 + n4)) : isNaN(r4) ? (t4(0, e2, s2 + i4), t4(2146959360, e2, s2 + n4)) : 17976931348623157e292 < r4 ? (t4(0, e2, s2 + i4), t4((u2 << 31 | 2146435072) >>> 0, e2, s2 + n4)) : r4 < 22250738585072014e-324 ? (t4((o2 = r4 / 5e-324) >>> 0, e2, s2 + i4), t4((u2 << 31 | o2 / 4294967296) >>> 0, e2, s2 + n4)) : (t4(4503599627370496 * (o2 = r4 * Math.pow(2, -(r4 = 1024 === (r4 = Math.floor(Math.log(r4) / Math.LN2)) ? 1023 : r4))) >>> 0, e2, s2 + i4), t4((u2 << 31 | r4 + 1023 << 20 | 1048576 * o2 & 1048575) >>> 0, e2, s2 + n4));
		      }
		      function d(t4, i4, n4, r4, e2) {
		        i4 = t4(r4, e2 + i4), t4 = t4(r4, e2 + n4), r4 = 2 * (t4 >> 31) + 1, e2 = t4 >>> 20 & 2047, n4 = 4294967296 * (1048575 & t4) + i4;
		        return 2047 == e2 ? n4 ? NaN : 1 / 0 * r4 : 0 == e2 ? 5e-324 * r4 * n4 : r4 * Math.pow(2, e2 - 1075) * (n4 + 4503599627370496);
		      }
		      function p(t4, i4, n4) {
		        h[0] = t4, i4[n4] = a[0], i4[n4 + 1] = a[1], i4[n4 + 2] = a[2], i4[n4 + 3] = a[3], i4[n4 + 4] = a[4], i4[n4 + 5] = a[5], i4[n4 + 6] = a[6], i4[n4 + 7] = a[7];
		      }
		      function v(t4, i4, n4) {
		        h[0] = t4, i4[n4] = a[7], i4[n4 + 1] = a[6], i4[n4 + 2] = a[5], i4[n4 + 3] = a[4], i4[n4 + 4] = a[3], i4[n4 + 5] = a[2], i4[n4 + 6] = a[1], i4[n4 + 7] = a[0];
		      }
		      function b(t4, i4) {
		        return a[0] = t4[i4], a[1] = t4[i4 + 1], a[2] = t4[i4 + 2], a[3] = t4[i4 + 3], a[4] = t4[i4 + 4], a[5] = t4[i4 + 5], a[6] = t4[i4 + 6], a[7] = t4[i4 + 7], h[0];
		      }
		      function w(t4, i4) {
		        return a[7] = t4[i4], a[6] = t4[i4 + 1], a[5] = t4[i4 + 2], a[4] = t4[i4 + 3], a[3] = t4[i4 + 4], a[2] = t4[i4 + 5], a[1] = t4[i4 + 6], a[0] = t4[i4 + 7], h[0];
		      }
		      return "undefined" != typeof Float32Array ? (u = new Float32Array([-0]), f = new Uint8Array(u.buffer), c = 128 === f[3], t3.writeFloatLE = c ? r3 : e, t3.writeFloatBE = c ? e : r3, t3.readFloatLE = c ? s : o, t3.readFloatBE = c ? o : s) : (t3.writeFloatLE = i3.bind(null, y), t3.writeFloatBE = i3.bind(null, m), t3.readFloatLE = n3.bind(null, g), t3.readFloatBE = n3.bind(null, j)), "undefined" != typeof Float64Array ? (h = new Float64Array([-0]), a = new Uint8Array(h.buffer), c = 128 === a[7], t3.writeDoubleLE = c ? p : v, t3.writeDoubleBE = c ? v : p, t3.readDoubleLE = c ? b : w, t3.readDoubleBE = c ? w : b) : (t3.writeDoubleLE = l.bind(null, y, 0, 4), t3.writeDoubleBE = l.bind(null, m, 4, 0), t3.readDoubleLE = d.bind(null, g, 0, 4), t3.readDoubleBE = d.bind(null, j, 4, 0)), t3;
		    }
		    function y(t3, i3, n3) {
		      i3[n3] = 255 & t3, i3[n3 + 1] = t3 >>> 8 & 255, i3[n3 + 2] = t3 >>> 16 & 255, i3[n3 + 3] = t3 >>> 24;
		    }
		    function m(t3, i3, n3) {
		      i3[n3] = t3 >>> 24, i3[n3 + 1] = t3 >>> 16 & 255, i3[n3 + 2] = t3 >>> 8 & 255, i3[n3 + 3] = 255 & t3;
		    }
		    function g(t3, i3) {
		      return (t3[i3] | t3[i3 + 1] << 8 | t3[i3 + 2] << 16 | t3[i3 + 3] << 24) >>> 0;
		    }
		    function j(t3, i3) {
		      return (t3[i3] << 24 | t3[i3 + 1] << 16 | t3[i3 + 2] << 8 | t3[i3 + 3]) >>> 0;
		    }
		    i2.exports = r2(r2);
		  }, {}], 7: [function(t, i, n) {
		    function r(t) {
		      try {
		        var i = eval("require")(t);
		        if (i && (i.length || Object.keys(i).length)) return i;
		      } catch (t2) {
		      }
		      return null;
		    }
		    i.exports = r;
		  }, {}], 8: [function(t2, i2, n2) {
		    var e = n2.isAbsolute = function(t3) {
		      return /^(?:\/|\w+:)/.test(t3);
		    }, r2 = n2.normalize = function(t3) {
		      var i3 = (t3 = t3.replace(/\\/g, "/").replace(/\/{2,}/g, "/")).split("/"), n3 = e(t3), t3 = "";
		      n3 && (t3 = i3.shift() + "/");
		      for (var r3 = 0; r3 < i3.length; ) ".." === i3[r3] ? 0 < r3 && ".." !== i3[r3 - 1] ? i3.splice(--r3, 2) : n3 ? i3.splice(r3, 1) : ++r3 : "." === i3[r3] ? i3.splice(r3, 1) : ++r3;
		      return t3 + i3.join("/");
		    };
		    n2.resolve = function(t3, i3, n3) {
		      return n3 || (i3 = r2(i3)), !e(i3) && (t3 = (t3 = n3 ? t3 : r2(t3)).replace(/(?:\/|^)[^/]+$/, "")).length ? r2(t3 + "/" + i3) : i3;
		    };
		  }, {}], 9: [function(t2, i2, n2) {
		    i2.exports = function(i3, n3, t3) {
		      var r2 = t3 || 8192, e = r2 >>> 1, s = null, o = r2;
		      return function(t4) {
		        if (t4 < 1 || e < t4) return i3(t4);
		        r2 < o + t4 && (s = i3(r2), o = 0);
		        t4 = n3.call(s, o, o += t4);
		        return 7 & o && (o = 1 + (7 | o)), t4;
		      };
		    };
		  }, {}], 10: [function(t2, i2, n2) {
		    n2.length = function(t3) {
		      for (var i3, n3 = 0, r2 = 0; r2 < t3.length; ++r2) (i3 = t3.charCodeAt(r2)) < 128 ? n3 += 1 : i3 < 2048 ? n3 += 2 : 55296 == (64512 & i3) && 56320 == (64512 & t3.charCodeAt(r2 + 1)) ? (++r2, n3 += 4) : n3 += 3;
		      return n3;
		    }, n2.read = function(t3, i3, n3) {
		      if (n3 - i3 < 1) return "";
		      for (var r2, e = null, s = [], o = 0; i3 < n3; ) (r2 = t3[i3++]) < 128 ? s[o++] = r2 : 191 < r2 && r2 < 224 ? s[o++] = (31 & r2) << 6 | 63 & t3[i3++] : 239 < r2 && r2 < 365 ? (r2 = ((7 & r2) << 18 | (63 & t3[i3++]) << 12 | (63 & t3[i3++]) << 6 | 63 & t3[i3++]) - 65536, s[o++] = 55296 + (r2 >> 10), s[o++] = 56320 + (1023 & r2)) : s[o++] = (15 & r2) << 12 | (63 & t3[i3++]) << 6 | 63 & t3[i3++], 8191 < o && ((e = e || []).push(String.fromCharCode.apply(String, s)), o = 0);
		      return e ? (o && e.push(String.fromCharCode.apply(String, s.slice(0, o))), e.join("")) : String.fromCharCode.apply(String, s.slice(0, o));
		    }, n2.write = function(t3, i3, n3) {
		      for (var r2, e, s = n3, o = 0; o < t3.length; ++o) (r2 = t3.charCodeAt(o)) < 128 ? i3[n3++] = r2 : (r2 < 2048 ? i3[n3++] = r2 >> 6 | 192 : (55296 == (64512 & r2) && 56320 == (64512 & (e = t3.charCodeAt(o + 1))) ? (++o, i3[n3++] = (r2 = 65536 + ((1023 & r2) << 10) + (1023 & e)) >> 18 | 240, i3[n3++] = r2 >> 12 & 63 | 128) : i3[n3++] = r2 >> 12 | 224, i3[n3++] = r2 >> 6 & 63 | 128), i3[n3++] = 63 & r2 | 128);
		      return n3 - s;
		    };
		  }, {}], 11: [function(t2, i2, n2) {
		    i2.exports = e;
		    var r2 = /\/|\./;
		    function e(t3, i3) {
		      r2.test(t3) || (t3 = "google/protobuf/" + t3 + ".proto", i3 = { nested: { google: { nested: { protobuf: { nested: i3 } } } } }), e[t3] = i3;
		    }
		    e("any", { Any: { fields: { type_url: { type: "string", id: 1 }, value: { type: "bytes", id: 2 } } } }), e("duration", { Duration: i2 = { fields: { seconds: { type: "int64", id: 1 }, nanos: { type: "int32", id: 2 } } } }), e("timestamp", { Timestamp: i2 }), e("empty", { Empty: { fields: {} } }), e("struct", { Struct: { fields: { fields: { keyType: "string", type: "Value", id: 1 } } }, Value: { oneofs: { kind: { oneof: ["nullValue", "numberValue", "stringValue", "boolValue", "structValue", "listValue"] } }, fields: { nullValue: { type: "NullValue", id: 1 }, numberValue: { type: "double", id: 2 }, stringValue: { type: "string", id: 3 }, boolValue: { type: "bool", id: 4 }, structValue: { type: "Struct", id: 5 }, listValue: { type: "ListValue", id: 6 } } }, NullValue: { values: { NULL_VALUE: 0 } }, ListValue: { fields: { values: { rule: "repeated", type: "Value", id: 1 } } } }), e("wrappers", { DoubleValue: { fields: { value: { type: "double", id: 1 } } }, FloatValue: { fields: { value: { type: "float", id: 1 } } }, Int64Value: { fields: { value: { type: "int64", id: 1 } } }, UInt64Value: { fields: { value: { type: "uint64", id: 1 } } }, Int32Value: { fields: { value: { type: "int32", id: 1 } } }, UInt32Value: { fields: { value: { type: "uint32", id: 1 } } }, BoolValue: { fields: { value: { type: "bool", id: 1 } } }, StringValue: { fields: { value: { type: "string", id: 1 } } }, BytesValue: { fields: { value: { type: "bytes", id: 1 } } } }), e("field_mask", { FieldMask: { fields: { paths: { rule: "repeated", type: "string", id: 1 } } } }), e.get = function(t3) {
		      return e[t3] || null;
		    };
		  }, {}], 12: [function(t2, i2, n2) {
		    var l = t2(15), d = t2(37);
		    function o(t3, i3, n3, r2) {
		      var e = false;
		      if (i3.resolvedType) if (i3.resolvedType instanceof l) {
		        t3("switch(d%s){", r2);
		        for (var s = i3.resolvedType.values, o2 = Object.keys(s), u = 0; u < o2.length; ++u) s[o2[u]] !== i3.typeDefault || e || (t3("default:")('if(typeof(d%s)==="number"){m%s=d%s;break}', r2, r2, r2), i3.repeated || t3("break"), e = true), t3("case%j:", o2[u])("case %i:", s[o2[u]])("m%s=%j", r2, s[o2[u]])("break");
		        t3("}");
		      } else t3('if(typeof d%s!=="object")', r2)("throw TypeError(%j)", i3.fullName + ": object expected")("m%s=types[%i].fromObject(d%s)", r2, n3, r2);
		      else {
		        var f = false;
		        switch (i3.type) {
		          case "double":
		          case "float":
		            t3("m%s=Number(d%s)", r2, r2);
		            break;
		          case "uint32":
		          case "fixed32":
		            t3("m%s=d%s>>>0", r2, r2);
		            break;
		          case "int32":
		          case "sint32":
		          case "sfixed32":
		            t3("m%s=d%s|0", r2, r2);
		            break;
		          case "uint64":
		            f = true;
		          case "int64":
		          case "sint64":
		          case "fixed64":
		          case "sfixed64":
		            t3("if(util.Long)")("(m%s=util.Long.fromValue(d%s)).unsigned=%j", r2, r2, f)('else if(typeof d%s==="string")', r2)("m%s=parseInt(d%s,10)", r2, r2)('else if(typeof d%s==="number")', r2)("m%s=d%s", r2, r2)('else if(typeof d%s==="object")', r2)("m%s=new util.LongBits(d%s.low>>>0,d%s.high>>>0).toNumber(%s)", r2, r2, r2, f ? "true" : "");
		            break;
		          case "bytes":
		            t3('if(typeof d%s==="string")', r2)("util.base64.decode(d%s,m%s=util.newBuffer(util.base64.length(d%s)),0)", r2, r2, r2)("else if(d%s.length >= 0)", r2)("m%s=d%s", r2, r2);
		            break;
		          case "string":
		            t3("m%s=String(d%s)", r2, r2);
		            break;
		          case "bool":
		            t3("m%s=Boolean(d%s)", r2, r2);
		        }
		      }
		      return t3;
		    }
		    function p(t3, i3, n3, r2) {
		      if (i3.resolvedType) i3.resolvedType instanceof l ? t3("d%s=o.enums===String?(types[%i].values[m%s]===undefined?m%s:types[%i].values[m%s]):m%s", r2, n3, r2, r2, n3, r2, r2) : t3("d%s=types[%i].toObject(m%s,o)", r2, n3, r2);
		      else {
		        var e = false;
		        switch (i3.type) {
		          case "double":
		          case "float":
		            t3("d%s=o.json&&!isFinite(m%s)?String(m%s):m%s", r2, r2, r2, r2);
		            break;
		          case "uint64":
		            e = true;
		          case "int64":
		          case "sint64":
		          case "fixed64":
		          case "sfixed64":
		            t3('if(typeof m%s==="number")', r2)("d%s=o.longs===String?String(m%s):m%s", r2, r2, r2)("else")("d%s=o.longs===String?util.Long.prototype.toString.call(m%s):o.longs===Number?new util.LongBits(m%s.low>>>0,m%s.high>>>0).toNumber(%s):m%s", r2, r2, r2, r2, e ? "true" : "", r2);
		            break;
		          case "bytes":
		            t3("d%s=o.bytes===String?util.base64.encode(m%s,0,m%s.length):o.bytes===Array?Array.prototype.slice.call(m%s):m%s", r2, r2, r2, r2, r2);
		            break;
		          default:
		            t3("d%s=m%s", r2, r2);
		        }
		      }
		      return t3;
		    }
		    n2.fromObject = function(t3) {
		      var i3 = t3.fieldsArray, n3 = d.codegen(["d"], t3.name + "$fromObject")("if(d instanceof this.ctor)")("return d");
		      if (!i3.length) return n3("return new this.ctor");
		      n3("var m=new this.ctor");
		      for (var r2 = 0; r2 < i3.length; ++r2) {
		        var e = i3[r2].resolve(), s = d.safeProp(e.name);
		        e.map ? (n3("if(d%s){", s)('if(typeof d%s!=="object")', s)("throw TypeError(%j)", e.fullName + ": object expected")("m%s={}", s)("for(var ks=Object.keys(d%s),i=0;i<ks.length;++i){", s), o(n3, e, r2, s + "[ks[i]]")("}")("}")) : e.repeated ? (n3("if(d%s){", s)("if(!Array.isArray(d%s))", s)("throw TypeError(%j)", e.fullName + ": array expected")("m%s=[]", s)("for(var i=0;i<d%s.length;++i){", s), o(n3, e, r2, s + "[i]")("}")("}")) : (e.resolvedType instanceof l || n3("if(d%s!=null){", s), o(n3, e, r2, s), e.resolvedType instanceof l || n3("}"));
		      }
		      return n3("return m");
		    }, n2.toObject = function(t3) {
		      var i3 = t3.fieldsArray.slice().sort(d.compareFieldsById);
		      if (!i3.length) return d.codegen()("return {}");
		      for (var n3 = d.codegen(["m", "o"], t3.name + "$toObject")("if(!o)")("o={}")("var d={}"), r2 = [], e = [], s = [], o2 = 0; o2 < i3.length; ++o2) i3[o2].partOf || (i3[o2].resolve().repeated ? r2 : i3[o2].map ? e : s).push(i3[o2]);
		      if (r2.length) {
		        for (n3("if(o.arrays||o.defaults){"), o2 = 0; o2 < r2.length; ++o2) n3("d%s=[]", d.safeProp(r2[o2].name));
		        n3("}");
		      }
		      if (e.length) {
		        for (n3("if(o.objects||o.defaults){"), o2 = 0; o2 < e.length; ++o2) n3("d%s={}", d.safeProp(e[o2].name));
		        n3("}");
		      }
		      if (s.length) {
		        for (n3("if(o.defaults){"), o2 = 0; o2 < s.length; ++o2) {
		          var u, f = s[o2], h = d.safeProp(f.name);
		          f.resolvedType instanceof l ? n3("d%s=o.enums===String?%j:%j", h, f.resolvedType.valuesById[f.typeDefault], f.typeDefault) : f.long ? n3("if(util.Long){")("var n=new util.Long(%i,%i,%j)", f.typeDefault.low, f.typeDefault.high, f.typeDefault.unsigned)("d%s=o.longs===String?n.toString():o.longs===Number?n.toNumber():n", h)("}else")("d%s=o.longs===String?%j:%i", h, f.typeDefault.toString(), f.typeDefault.toNumber()) : f.bytes ? (u = "[" + Array.prototype.slice.call(f.typeDefault).join(",") + "]", n3("if(o.bytes===String)d%s=%j", h, String.fromCharCode.apply(String, f.typeDefault))("else{")("d%s=%s", h, u)("if(o.bytes!==Array)d%s=util.newBuffer(d%s)", h, h)("}")) : n3("d%s=%j", h, f.typeDefault);
		        }
		        n3("}");
		      }
		      for (var a = false, o2 = 0; o2 < i3.length; ++o2) {
		        var f = i3[o2], c = t3.e.indexOf(f), h = d.safeProp(f.name);
		        f.map ? (a || (a = true, n3("var ks2")), n3("if(m%s&&(ks2=Object.keys(m%s)).length){", h, h)("d%s={}", h)("for(var j=0;j<ks2.length;++j){"), p(n3, f, c, h + "[ks2[j]]")("}")) : f.repeated ? (n3("if(m%s&&m%s.length){", h, h)("d%s=[]", h)("for(var j=0;j<m%s.length;++j){", h), p(n3, f, c, h + "[j]")("}")) : (n3("if(m%s!=null&&m.hasOwnProperty(%j)){", h, f.name), p(n3, f, c, h), f.partOf && n3("if(o.oneofs)")("d%s=%j", d.safeProp(f.partOf.name), f.name)), n3("}");
		      }
		      return n3("return d");
		    };
		  }, { 15: 15, 37: 37 }], 13: [function(t2, i2, n2) {
		    i2.exports = function(t3) {
		      for (var i3 = h.codegen(["r", "l", "e"], t3.name + "$decode")("if(!(r instanceof Reader))")("r=Reader.create(r)")("var c=l===undefined?r.len:r.pos+l,m=new this.ctor" + (t3.fieldsArray.filter(function(t4) {
		        return t4.map;
		      }).length ? ",k,value" : ""))("while(r.pos<c){")("var t=r.uint32()")("if(t===e)")("break")("switch(t>>>3){"), n3 = 0; n3 < t3.fieldsArray.length; ++n3) {
		        var r2 = t3.e[n3].resolve(), e = r2.resolvedType instanceof u ? "int32" : r2.type, s = "m" + h.safeProp(r2.name);
		        i3("case %i: {", r2.id), r2.map ? (i3("if(%s===util.emptyObject)", s)("%s={}", s)("var c2 = r.uint32()+r.pos"), f.defaults[r2.keyType] !== it ? i3("k=%j", f.defaults[r2.keyType]) : i3("k=null"), f.defaults[e] !== it ? i3("value=%j", f.defaults[e]) : i3("value=null"), i3("while(r.pos<c2){")("var tag2=r.uint32()")("switch(tag2>>>3){")("case 1: k=r.%s(); break", r2.keyType)("case 2:"), f.basic[e] === it ? i3("value=types[%i].decode(r,r.uint32())", n3) : i3("value=r.%s()", e), i3("break")("default:")("r.skipType(tag2&7)")("break")("}")("}"), f.long[r2.keyType] !== it ? i3('%s[typeof k==="object"?util.longToHash(k):k]=value', s) : i3("%s[k]=value", s)) : r2.repeated ? (i3("if(!(%s&&%s.length))", s, s)("%s=[]", s), f.packed[e] !== it && i3("if((t&7)===2){")("var c2=r.uint32()+r.pos")("while(r.pos<c2)")("%s.push(r.%s())", s, e)("}else"), f.basic[e] === it ? i3(r2.delimited ? "%s.push(types[%i].decode(r,undefined,((t&~7)|4)))" : "%s.push(types[%i].decode(r,r.uint32()))", s, n3) : i3("%s.push(r.%s())", s, e)) : f.basic[e] === it ? i3(r2.delimited ? "%s=types[%i].decode(r,undefined,((t&~7)|4))" : "%s=types[%i].decode(r,r.uint32())", s, n3) : i3("%s=r.%s()", s, e), i3("break")("}");
		      }
		      for (i3("default:")("r.skipType(t&7)")("break")("}")("}"), n3 = 0; n3 < t3.e.length; ++n3) {
		        var o = t3.e[n3];
		        o.required && i3("if(!m.hasOwnProperty(%j))", o.name)("throw util.ProtocolError(%j,{instance:m})", "missing required '" + o.name + "'");
		      }
		      return i3("return m");
		    };
		    var u = t2(15), f = t2(36), h = t2(37);
		  }, { 15: 15, 36: 36, 37: 37 }], 14: [function(t2, i2, n2) {
		    i2.exports = function(t3) {
		      for (var i3, n3 = c.codegen(["m", "w"], t3.name + "$encode")("if(!w)")("w=Writer.create()"), r2 = t3.fieldsArray.slice().sort(c.compareFieldsById), e = 0; e < r2.length; ++e) {
		        var s = r2[e].resolve(), o = t3.e.indexOf(s), u = s.resolvedType instanceof h ? "int32" : s.type, f = a.basic[u];
		        i3 = "m" + c.safeProp(s.name), s.map ? (n3("if(%s!=null&&Object.hasOwnProperty.call(m,%j)){", i3, s.name)("for(var ks=Object.keys(%s),i=0;i<ks.length;++i){", i3)("w.uint32(%i).fork().uint32(%i).%s(ks[i])", (s.id << 3 | 2) >>> 0, 8 | a.mapKey[s.keyType], s.keyType), f === it ? n3("types[%i].encode(%s[ks[i]],w.uint32(18).fork()).ldelim().ldelim()", o, i3) : n3(".uint32(%i).%s(%s[ks[i]]).ldelim()", 16 | f, u, i3), n3("}")("}")) : s.repeated ? (n3("if(%s!=null&&%s.length){", i3, i3), s.packed && a.packed[u] !== it ? n3("w.uint32(%i).fork()", (s.id << 3 | 2) >>> 0)("for(var i=0;i<%s.length;++i)", i3)("w.%s(%s[i])", u, i3)("w.ldelim()") : (n3("for(var i=0;i<%s.length;++i)", i3), f === it ? l(n3, s, o, i3 + "[i]") : n3("w.uint32(%i).%s(%s[i])", (s.id << 3 | f) >>> 0, u, i3)), n3("}")) : (s.optional && n3("if(%s!=null&&Object.hasOwnProperty.call(m,%j))", i3, s.name), f === it ? l(n3, s, o, i3) : n3("w.uint32(%i).%s(%s)", (s.id << 3 | f) >>> 0, u, i3));
		      }
		      return n3("return w");
		    };
		    var h = t2(15), a = t2(36), c = t2(37);
		    function l(t3, i3, n3, r2) {
		      i3.delimited ? t3("types[%i].encode(%s,w.uint32(%i)).uint32(%i)", n3, r2, (i3.id << 3 | 3) >>> 0, (i3.id << 3 | 4) >>> 0) : t3("types[%i].encode(%s,w.uint32(%i).fork()).ldelim()", n3, r2, (i3.id << 3 | 2) >>> 0);
		    }
		  }, { 15: 15, 36: 36, 37: 37 }], 15: [function(t2, i2, n2) {
		    i2.exports = s;
		    var f = t2(24), r2 = (((s.prototype = Object.create(f.prototype)).constructor = s).className = "Enum", t2(23)), e = t2(37);
		    function s(t3, i3, n3, r3, e2, s2) {
		      if (f.call(this, t3, n3), i3 && "object" != typeof i3) throw TypeError("values must be an object");
		      if (this.valuesById = {}, this.values = Object.create(this.valuesById), this.comment = r3, this.comments = e2 || {}, this.valuesOptions = s2, this.o = {}, this.reserved = it, i3) for (var o = Object.keys(i3), u = 0; u < o.length; ++u) "number" == typeof i3[o[u]] && (this.valuesById[this.values[o[u]] = i3[o[u]]] = o[u]);
		    }
		    s.prototype.u = function(t3) {
		      return t3 = this.f || t3, f.prototype.u.call(this, t3), Object.keys(this.values).forEach((t4) => {
		        var i3 = Object.assign({}, this.h);
		        this.o[t4] = Object.assign(i3, this.valuesOptions && this.valuesOptions[t4] && this.valuesOptions[t4].features);
		      }), this;
		    }, s.fromJSON = function(t3, i3) {
		      t3 = new s(t3, i3.values, i3.options, i3.comment, i3.comments);
		      return t3.reserved = i3.reserved, i3.edition && (t3.f = i3.edition), t3.a = "proto3", t3;
		    }, s.prototype.toJSON = function(t3) {
		      t3 = !!t3 && !!t3.keepComments;
		      return e.toObject(["edition", this.c(), "options", this.options, "valuesOptions", this.valuesOptions, "values", this.values, "reserved", this.reserved && this.reserved.length ? this.reserved : it, "comment", t3 ? this.comment : it, "comments", t3 ? this.comments : it]);
		    }, s.prototype.add = function(t3, i3, n3, r3) {
		      if (!e.isString(t3)) throw TypeError("name must be a string");
		      if (!e.isInteger(i3)) throw TypeError("id must be an integer");
		      if (this.values[t3] !== it) throw Error("duplicate name '" + t3 + "' in " + this);
		      if (this.isReservedId(i3)) throw Error("id " + i3 + " is reserved in " + this);
		      if (this.isReservedName(t3)) throw Error("name '" + t3 + "' is reserved in " + this);
		      if (this.valuesById[i3] !== it) {
		        if (!this.options || !this.options.allow_alias) throw Error("duplicate id " + i3 + " in " + this);
		        this.values[t3] = i3;
		      } else this.valuesById[this.values[t3] = i3] = t3;
		      return r3 && (this.valuesOptions === it && (this.valuesOptions = {}), this.valuesOptions[t3] = r3 || null), this.comments[t3] = n3 || null, this;
		    }, s.prototype.remove = function(t3) {
		      if (!e.isString(t3)) throw TypeError("name must be a string");
		      var i3 = this.values[t3];
		      if (null == i3) throw Error("name '" + t3 + "' does not exist in " + this);
		      return delete this.valuesById[i3], delete this.values[t3], delete this.comments[t3], this.valuesOptions && delete this.valuesOptions[t3], this;
		    }, s.prototype.isReservedId = function(t3) {
		      return r2.isReservedId(this.reserved, t3);
		    }, s.prototype.isReservedName = function(t3) {
		      return r2.isReservedName(this.reserved, t3);
		    };
		  }, { 23: 23, 24: 24, 37: 37 }], 16: [function(t2, i2, n2) {
		    i2.exports = o;
		    var r2, u = t2(24), e = (((o.prototype = Object.create(u.prototype)).constructor = o).className = "Field", t2(15)), f = t2(36), h = t2(37), a = /^required|optional|repeated$/;
		    function o(t3, i3, n3, r3, e2, s, o2) {
		      if (h.isObject(r3) ? (o2 = e2, s = r3, r3 = e2 = it) : h.isObject(e2) && (o2 = s, s = e2, e2 = it), u.call(this, t3, s), !h.isInteger(i3) || i3 < 0) throw TypeError("id must be a non-negative integer");
		      if (!h.isString(n3)) throw TypeError("type must be a string");
		      if (r3 !== it && !a.test(r3 = r3.toString().toLowerCase())) throw TypeError("rule must be a string rule");
		      if (e2 !== it && !h.isString(e2)) throw TypeError("extend must be a string");
		      this.rule = (r3 = "proto3_optional" === r3 ? "optional" : r3) && "optional" !== r3 ? r3 : it, this.type = n3, this.id = i3, this.extend = e2 || it, this.repeated = "repeated" === r3, this.map = false, this.message = null, this.partOf = null, this.typeDefault = null, this.defaultValue = null, this.long = !!h.Long && f.long[n3] !== it, this.bytes = "bytes" === n3, this.resolvedType = null, this.extensionField = null, this.declaringField = null, this.comment = o2;
		    }
		    o.fromJSON = function(t3, i3) {
		      t3 = new o(t3, i3.id, i3.type, i3.rule, i3.extend, i3.options, i3.comment);
		      return i3.edition && (t3.f = i3.edition), t3.a = "proto3", t3;
		    }, Object.defineProperty(o.prototype, "required", { get: function() {
		      return "LEGACY_REQUIRED" === this.h.field_presence;
		    } }), Object.defineProperty(o.prototype, "optional", { get: function() {
		      return !this.required;
		    } }), Object.defineProperty(o.prototype, "delimited", { get: function() {
		      return this.resolvedType instanceof r2 && "DELIMITED" === this.h.message_encoding;
		    } }), Object.defineProperty(o.prototype, "packed", { get: function() {
		      return "PACKED" === this.h.repeated_field_encoding;
		    } }), Object.defineProperty(o.prototype, "hasPresence", { get: function() {
		      return !this.repeated && !this.map && (this.partOf || this.declaringField || this.extensionField || "IMPLICIT" !== this.h.field_presence);
		    } }), o.prototype.setOption = function(t3, i3, n3) {
		      return u.prototype.setOption.call(this, t3, i3, n3);
		    }, o.prototype.toJSON = function(t3) {
		      t3 = !!t3 && !!t3.keepComments;
		      return h.toObject(["edition", this.c(), "rule", "optional" !== this.rule && this.rule || it, "type", this.type, "id", this.id, "extend", this.extend, "options", this.options, "comment", t3 ? this.comment : it]);
		    }, o.prototype.resolve = function() {
		      var t3;
		      return this.resolved ? this : ((this.typeDefault = f.defaults[this.type]) === it ? (this.resolvedType = (this.declaringField || this).parent.lookupTypeOrEnum(this.type), this.resolvedType instanceof r2 ? this.typeDefault = null : this.typeDefault = this.resolvedType.values[Object.keys(this.resolvedType.values)[0]]) : this.options && this.options.proto3_optional && (this.typeDefault = null), this.options && null != this.options.default && (this.typeDefault = this.options.default, this.resolvedType instanceof e && "string" == typeof this.typeDefault && (this.typeDefault = this.resolvedType.values[this.typeDefault])), this.options && (this.options.packed === it || !this.resolvedType || this.resolvedType instanceof e || delete this.options.packed, Object.keys(this.options).length || (this.options = it)), this.long ? (this.typeDefault = h.Long.fromNumber(this.typeDefault, "u" == (this.type[0] || "")), Object.freeze && Object.freeze(this.typeDefault)) : this.bytes && "string" == typeof this.typeDefault && (h.base64.test(this.typeDefault) ? h.base64.decode(this.typeDefault, t3 = h.newBuffer(h.base64.length(this.typeDefault)), 0) : h.utf8.write(this.typeDefault, t3 = h.newBuffer(h.utf8.length(this.typeDefault)), 0), this.typeDefault = t3), this.map ? this.defaultValue = h.emptyObject : this.repeated ? this.defaultValue = h.emptyArray : this.defaultValue = this.typeDefault, this.parent instanceof r2 && (this.parent.ctor.prototype[this.name] = this.defaultValue), u.prototype.resolve.call(this));
		    }, o.prototype.l = function(t3) {
		      var i3;
		      return "proto2" !== t3 && "proto3" !== t3 ? {} : (t3 = {}, "required" === this.rule && (t3.field_presence = "LEGACY_REQUIRED"), this.parent && f.defaults[this.type] === it && (i3 = this.parent.get(this.type.split(".").pop())) && i3 instanceof r2 && i3.group && (t3.message_encoding = "DELIMITED"), true === this.getOption("packed") ? t3.repeated_field_encoding = "PACKED" : false === this.getOption("packed") && (t3.repeated_field_encoding = "EXPANDED"), t3);
		    }, o.prototype.u = function(t3) {
		      return u.prototype.u.call(this, this.f || t3);
		    }, o.d = function(n3, r3, e2, s) {
		      return "function" == typeof r3 ? r3 = h.decorateType(r3).name : r3 && "object" == typeof r3 && (r3 = h.decorateEnum(r3).name), function(t3, i3) {
		        h.decorateType(t3.constructor).add(new o(i3, n3, r3, e2, { default: s }));
		      };
		    }, o.p = function(t3) {
		      r2 = t3;
		    };
		  }, { 15: 15, 24: 24, 36: 36, 37: 37 }], 17: [function(t2, i2, n2) {
		    var r2 = i2.exports = t2(18);
		    r2.build = "light", r2.load = function(t3, i3, n3) {
		      return (i3 = "function" == typeof i3 ? (n3 = i3, new r2.Root()) : i3 || new r2.Root()).load(t3, n3);
		    }, r2.loadSync = function(t3, i3) {
		      return (i3 = i3 || new r2.Root()).loadSync(t3);
		    }, r2.encoder = t2(14), r2.decoder = t2(13), r2.verifier = t2(40), r2.converter = t2(12), r2.ReflectionObject = t2(24), r2.Namespace = t2(23), r2.Root = t2(29), r2.Enum = t2(15), r2.Type = t2(35), r2.Field = t2(16), r2.OneOf = t2(25), r2.MapField = t2(20), r2.Service = t2(33), r2.Method = t2(22), r2.Message = t2(21), r2.wrappers = t2(41), r2.types = t2(36), r2.util = t2(37), r2.ReflectionObject.p(r2.Root), r2.Namespace.p(r2.Type, r2.Service, r2.Enum), r2.Root.p(r2.Type), r2.Field.p(r2.Type);
		  }, { 12: 12, 13: 13, 14: 14, 15: 15, 16: 16, 18: 18, 20: 20, 21: 21, 22: 22, 23: 23, 24: 24, 25: 25, 29: 29, 33: 33, 35: 35, 36: 36, 37: 37, 40: 40, 41: 41 }], 18: [function(t2, i2, n2) {
		    var r2 = n2;
		    function e() {
		      r2.util.p(), r2.Writer.p(r2.BufferWriter), r2.Reader.p(r2.BufferReader);
		    }
		    r2.build = "minimal", r2.Writer = t2(42), r2.BufferWriter = t2(43), r2.Reader = t2(27), r2.BufferReader = t2(28), r2.util = t2(39), r2.rpc = t2(31), r2.roots = t2(30), r2.configure = e, e();
		  }, { 27: 27, 28: 28, 30: 30, 31: 31, 39: 39, 42: 42, 43: 43 }], 19: [function(t2, i2, n2) {
		    i2 = i2.exports = t2(17);
		    i2.build = "full", i2.tokenize = t2(34), i2.parse = t2(26), i2.common = t2(11), i2.Root.p(i2.Type, i2.parse, i2.common);
		  }, { 11: 11, 17: 17, 26: 26, 34: 34 }], 20: [function(t2, i2, n2) {
		    i2.exports = s;
		    var o = t2(16), r2 = (((s.prototype = Object.create(o.prototype)).constructor = s).className = "MapField", t2(36)), u = t2(37);
		    function s(t3, i3, n3, r3, e, s2) {
		      if (o.call(this, t3, i3, r3, it, it, e, s2), !u.isString(n3)) throw TypeError("keyType must be a string");
		      this.keyType = n3, this.resolvedKeyType = null, this.map = true;
		    }
		    s.fromJSON = function(t3, i3) {
		      return new s(t3, i3.id, i3.keyType, i3.type, i3.options, i3.comment);
		    }, s.prototype.toJSON = function(t3) {
		      t3 = !!t3 && !!t3.keepComments;
		      return u.toObject(["keyType", this.keyType, "type", this.type, "id", this.id, "extend", this.extend, "options", this.options, "comment", t3 ? this.comment : it]);
		    }, s.prototype.resolve = function() {
		      if (this.resolved) return this;
		      if (r2.mapKey[this.keyType] === it) throw Error("invalid key type: " + this.keyType);
		      return o.prototype.resolve.call(this);
		    }, s.d = function(n3, r3, e) {
		      return "function" == typeof e ? e = u.decorateType(e).name : e && "object" == typeof e && (e = u.decorateEnum(e).name), function(t3, i3) {
		        u.decorateType(t3.constructor).add(new s(i3, n3, r3, e));
		      };
		    };
		  }, { 16: 16, 36: 36, 37: 37 }], 21: [function(t2, i2, n2) {
		    i2.exports = e;
		    var r2 = t2(39);
		    function e(t3) {
		      if (t3) for (var i3 = Object.keys(t3), n3 = 0; n3 < i3.length; ++n3) this[i3[n3]] = t3[i3[n3]];
		    }
		    e.create = function(t3) {
		      return this.$type.create(t3);
		    }, e.encode = function(t3, i3) {
		      return this.$type.encode(t3, i3);
		    }, e.encodeDelimited = function(t3, i3) {
		      return this.$type.encodeDelimited(t3, i3);
		    }, e.decode = function(t3) {
		      return this.$type.decode(t3);
		    }, e.decodeDelimited = function(t3) {
		      return this.$type.decodeDelimited(t3);
		    }, e.verify = function(t3) {
		      return this.$type.verify(t3);
		    }, e.fromObject = function(t3) {
		      return this.$type.fromObject(t3);
		    }, e.toObject = function(t3, i3) {
		      return this.$type.toObject(t3, i3);
		    }, e.prototype.toJSON = function() {
		      return this.$type.toObject(this, r2.toJSONOptions);
		    };
		  }, { 39: 39 }], 22: [function(t2, i2, n2) {
		    i2.exports = r2;
		    var h = t2(24), a = (((r2.prototype = Object.create(h.prototype)).constructor = r2).className = "Method", t2(37));
		    function r2(t3, i3, n3, r3, e, s, o, u, f) {
		      if (a.isObject(e) ? (o = e, e = s = it) : a.isObject(s) && (o = s, s = it), i3 !== it && !a.isString(i3)) throw TypeError("type must be a string");
		      if (!a.isString(n3)) throw TypeError("requestType must be a string");
		      if (!a.isString(r3)) throw TypeError("responseType must be a string");
		      h.call(this, t3, o), this.type = i3 || "rpc", this.requestType = n3, this.requestStream = !!e || it, this.responseType = r3, this.responseStream = !!s || it, this.resolvedRequestType = null, this.resolvedResponseType = null, this.comment = u, this.parsedOptions = f;
		    }
		    r2.fromJSON = function(t3, i3) {
		      return new r2(t3, i3.type, i3.requestType, i3.responseType, i3.requestStream, i3.responseStream, i3.options, i3.comment, i3.parsedOptions);
		    }, r2.prototype.toJSON = function(t3) {
		      t3 = !!t3 && !!t3.keepComments;
		      return a.toObject(["type", "rpc" !== this.type && this.type || it, "requestType", this.requestType, "requestStream", this.requestStream, "responseType", this.responseType, "responseStream", this.responseStream, "options", this.options, "comment", t3 ? this.comment : it, "parsedOptions", this.parsedOptions]);
		    }, r2.prototype.resolve = function() {
		      return this.resolved ? this : (this.resolvedRequestType = this.parent.lookupType(this.requestType), this.resolvedResponseType = this.parent.lookupType(this.responseType), h.prototype.resolve.call(this));
		    };
		  }, { 24: 24, 37: 37 }], 23: [function(t2, i2, n2) {
		    i2.exports = c;
		    var s, o, u, r2 = t2(24), f = (((c.prototype = Object.create(r2.prototype)).constructor = c).className = "Namespace", t2(16)), h = t2(37), a = t2(25);
		    function e(t3, i3) {
		      if (!t3 || !t3.length) return it;
		      for (var n3 = {}, r3 = 0; r3 < t3.length; ++r3) n3[t3[r3].name] = t3[r3].toJSON(i3);
		      return n3;
		    }
		    function c(t3, i3) {
		      r2.call(this, t3, i3), this.nested = it, this.v = null, this.b = {}, this.w = true, this.y = true;
		    }
		    function l(t3) {
		      t3.v = null, t3.b = {};
		      for (var i3 = t3; i3 = i3.parent; ) i3.b = {};
		      return t3;
		    }
		    c.fromJSON = function(t3, i3) {
		      return new c(t3, i3.options).addJSON(i3.nested);
		    }, c.arrayToJSON = e, c.isReservedId = function(t3, i3) {
		      if (t3) {
		        for (var n3 = 0; n3 < t3.length; ++n3) if ("string" != typeof t3[n3] && t3[n3][0] <= i3 && t3[n3][1] > i3) return true;
		      }
		      return false;
		    }, c.isReservedName = function(t3, i3) {
		      if (t3) {
		        for (var n3 = 0; n3 < t3.length; ++n3) if (t3[n3] === i3) return true;
		      }
		      return false;
		    }, Object.defineProperty(c.prototype, "nestedArray", { get: function() {
		      return this.v || (this.v = h.toArray(this.nested));
		    } }), c.prototype.toJSON = function(t3) {
		      return h.toObject(["options", this.options, "nested", e(this.nestedArray, t3)]);
		    }, c.prototype.addJSON = function(t3) {
		      if (t3) for (var i3, n3 = Object.keys(t3), r3 = 0; r3 < n3.length; ++r3) i3 = t3[n3[r3]], this.add((i3.fields !== it ? s : i3.values !== it ? u : i3.methods !== it ? o : i3.id !== it ? f : c).fromJSON(n3[r3], i3));
		      return this;
		    }, c.prototype.get = function(t3) {
		      return this.nested && this.nested[t3] || null;
		    }, c.prototype.getEnum = function(t3) {
		      if (this.nested && this.nested[t3] instanceof u) return this.nested[t3].values;
		      throw Error("no such enum: " + t3);
		    }, c.prototype.add = function(t3) {
		      if (!(t3 instanceof f && t3.extend !== it || t3 instanceof s || t3 instanceof a || t3 instanceof u || t3 instanceof o || t3 instanceof c)) throw TypeError("object must be a valid nested object");
		      if (this.nested) {
		        var i3 = this.get(t3.name);
		        if (i3) {
		          if (!(i3 instanceof c && t3 instanceof c) || i3 instanceof s || i3 instanceof o) throw Error("duplicate name '" + t3.name + "' in " + this);
		          for (var n3 = i3.nestedArray, r3 = 0; r3 < n3.length; ++r3) t3.add(n3[r3]);
		          this.remove(i3), this.nested || (this.nested = {}), t3.setOptions(i3.options, true);
		        }
		      } else this.nested = {};
		      this.nested[t3.name] = t3, this instanceof s || this instanceof o || this instanceof u || this instanceof f || t3.f || (t3.f = t3.a), this.w = true, this.y = true;
		      for (var e2 = this; e2 = e2.parent; ) e2.w = true, e2.y = true;
		      return t3.onAdd(this), l(this);
		    }, c.prototype.remove = function(t3) {
		      if (!(t3 instanceof r2)) throw TypeError("object must be a ReflectionObject");
		      if (t3.parent !== this) throw Error(t3 + " is not a member of " + this);
		      return delete this.nested[t3.name], Object.keys(this.nested).length || (this.nested = it), t3.onRemove(this), l(this);
		    }, c.prototype.define = function(t3, i3) {
		      if (h.isString(t3)) t3 = t3.split(".");
		      else if (!Array.isArray(t3)) throw TypeError("illegal path");
		      if (t3 && t3.length && "" === t3[0]) throw Error("path must be relative");
		      for (var n3 = this; 0 < t3.length; ) {
		        var r3 = t3.shift();
		        if (n3.nested && n3.nested[r3]) {
		          if (!((n3 = n3.nested[r3]) instanceof c)) throw Error("path conflicts with non-namespace objects");
		        } else n3.add(n3 = new c(r3));
		      }
		      return i3 && n3.addJSON(i3), n3;
		    }, c.prototype.resolveAll = function() {
		      if (this.y) {
		        this.g(this.f);
		        var t3 = this.nestedArray, i3 = 0;
		        for (this.resolve(); i3 < t3.length; ) t3[i3] instanceof c ? t3[i3++].resolveAll() : t3[i3++].resolve();
		        this.y = false;
		      }
		      return this;
		    }, c.prototype.g = function(i3) {
		      return this.w && (this.w = false, i3 = this.f || i3, r2.prototype.g.call(this, i3), this.nestedArray.forEach((t3) => {
		        t3.g(i3);
		      })), this;
		    }, c.prototype.lookup = function(t3, i3, n3) {
		      if ("boolean" == typeof i3 ? (n3 = i3, i3 = it) : i3 && !Array.isArray(i3) && (i3 = [i3]), h.isString(t3) && t3.length) {
		        if ("." === t3) return this.root;
		        t3 = t3.split(".");
		      } else if (!t3.length) return this;
		      var r3 = t3.join(".");
		      if ("" === t3[0]) return this.root.lookup(t3.slice(1), i3);
		      var e2 = this.root.j && this.root.j["." + r3];
		      if (e2 && (!i3 || ~i3.indexOf(e2.constructor))) return e2;
		      if ((e2 = this.k(t3, r3)) && (!i3 || ~i3.indexOf(e2.constructor))) return e2;
		      if (!n3) for (var s2 = this; s2.parent; ) {
		        if ((e2 = s2.parent.k(t3, r3)) && (!i3 || ~i3.indexOf(e2.constructor))) return e2;
		        s2 = s2.parent;
		      }
		      return null;
		    }, c.prototype.k = function(t3, i3) {
		      if (Object.prototype.hasOwnProperty.call(this.b, i3)) return this.b[i3];
		      var n3 = this.get(t3[0]), r3 = null;
		      if (n3) 1 === t3.length ? r3 = n3 : n3 instanceof c && (t3 = t3.slice(1), r3 = n3.k(t3, t3.join(".")));
		      else for (var e2 = 0; e2 < this.nestedArray.length; ++e2) this.v[e2] instanceof c && (n3 = this.v[e2].k(t3, i3)) && (r3 = n3);
		      return this.b[i3] = r3;
		    }, c.prototype.lookupType = function(t3) {
		      var i3 = this.lookup(t3, [s]);
		      if (i3) return i3;
		      throw Error("no such type: " + t3);
		    }, c.prototype.lookupEnum = function(t3) {
		      var i3 = this.lookup(t3, [u]);
		      if (i3) return i3;
		      throw Error("no such Enum '" + t3 + "' in " + this);
		    }, c.prototype.lookupTypeOrEnum = function(t3) {
		      var i3 = this.lookup(t3, [s, u]);
		      if (i3) return i3;
		      throw Error("no such Type or Enum '" + t3 + "' in " + this);
		    }, c.prototype.lookupService = function(t3) {
		      var i3 = this.lookup(t3, [o]);
		      if (i3) return i3;
		      throw Error("no such Service '" + t3 + "' in " + this);
		    }, c.p = function(t3, i3, n3) {
		      s = t3, o = i3, u = n3;
		    };
		  }, { 16: 16, 24: 24, 25: 25, 37: 37 }], 24: [function(t2, i2, n2) {
		    (i2.exports = a).className = "ReflectionObject";
		    const r2 = t2(25);
		    var e, o = t2(37), s = { enum_type: "OPEN", field_presence: "EXPLICIT", json_format: "ALLOW", message_encoding: "LENGTH_PREFIXED", repeated_field_encoding: "PACKED", utf8_validation: "VERIFY", enforce_naming_style: "STYLE2024", default_symbol_visibility: "EXPORT_TOP_LEVEL" }, u = { enum_type: "OPEN", field_presence: "EXPLICIT", json_format: "ALLOW", message_encoding: "LENGTH_PREFIXED", repeated_field_encoding: "PACKED", utf8_validation: "VERIFY", enforce_naming_style: "STYLE_LEGACY", default_symbol_visibility: "EXPORT_ALL" }, f = { enum_type: "CLOSED", field_presence: "EXPLICIT", json_format: "LEGACY_BEST_EFFORT", message_encoding: "LENGTH_PREFIXED", repeated_field_encoding: "EXPANDED", utf8_validation: "NONE", enforce_naming_style: "STYLE_LEGACY", default_symbol_visibility: "EXPORT_ALL" }, h = { enum_type: "OPEN", field_presence: "IMPLICIT", json_format: "ALLOW", message_encoding: "LENGTH_PREFIXED", repeated_field_encoding: "PACKED", utf8_validation: "VERIFY", enforce_naming_style: "STYLE_LEGACY", default_symbol_visibility: "EXPORT_ALL" };
		    function a(t3, i3) {
		      if (!o.isString(t3)) throw TypeError("name must be a string");
		      if (i3 && !o.isObject(i3)) throw TypeError("options must be an object");
		      this.options = i3, this.parsedOptions = null, this.name = t3, this.f = null, this.a = "proto2", this.h = {}, this.O = false, this.parent = null, this.resolved = false, this.comment = null, this.filename = null;
		    }
		    Object.defineProperties(a.prototype, { root: { get: function() {
		      for (var t3 = this; null !== t3.parent; ) t3 = t3.parent;
		      return t3;
		    } }, fullName: { get: function() {
		      for (var t3 = [this.name], i3 = this.parent; i3; ) t3.unshift(i3.name), i3 = i3.parent;
		      return t3.join(".");
		    } } }), a.prototype.toJSON = function() {
		      throw Error();
		    }, a.prototype.onAdd = function(t3) {
		      this.parent && this.parent !== t3 && this.parent.remove(this), this.parent = t3, this.resolved = false;
		      t3 = t3.root;
		      t3 instanceof e && t3._(this);
		    }, a.prototype.onRemove = function(t3) {
		      t3 = t3.root;
		      t3 instanceof e && t3.A(this), this.parent = null, this.resolved = false;
		    }, a.prototype.resolve = function() {
		      return this.resolved || this.root instanceof e && (this.resolved = true), this;
		    }, a.prototype.g = function(t3) {
		      return this.u(this.f || t3);
		    }, a.prototype.u = function(t3) {
		      if (!this.O) {
		        var i3 = {};
		        if (!t3) throw Error("Unknown edition for " + this.fullName);
		        var n3 = Object.assign(this.options ? Object.assign({}, this.options.features) : {}, this.l(t3));
		        if (this.f) {
		          if ("proto2" === t3) i3 = Object.assign({}, f);
		          else if ("proto3" === t3) i3 = Object.assign({}, h);
		          else if ("2023" === t3) i3 = Object.assign({}, u);
		          else {
		            if ("2024" !== t3) throw Error("Unknown edition: " + t3);
		            i3 = Object.assign({}, s);
		          }
		          this.h = Object.assign(i3, n3 || {});
		        } else {
		          if (this.partOf instanceof r2) {
		            t3 = Object.assign({}, this.partOf.h);
		            this.h = Object.assign(t3, n3 || {});
		          } else if (!this.declaringField) {
		            if (!this.parent) throw Error("Unable to find a parent for " + this.fullName);
		            i3 = Object.assign({}, this.parent.h);
		            this.h = Object.assign(i3, n3 || {});
		          }
		          this.extensionField && (this.extensionField.h = this.h);
		        }
		        this.O = true;
		      }
		    }, a.prototype.l = function() {
		      return {};
		    }, a.prototype.getOption = function(t3) {
		      return this.options ? this.options[t3] : it;
		    }, a.prototype.setOption = function(t3, i3, n3) {
		      return this.options || (this.options = {}), /^features\./.test(t3) ? o.setProperty(this.options, t3, i3, n3) : n3 && this.options[t3] !== it || (this.getOption(t3) !== i3 && (this.resolved = false), this.options[t3] = i3), this;
		    }, a.prototype.setParsedOption = function(i3, t3, n3) {
		      this.parsedOptions || (this.parsedOptions = []);
		      var r3, e2, s2 = this.parsedOptions;
		      return n3 ? (r3 = s2.find(function(t4) {
		        return Object.prototype.hasOwnProperty.call(t4, i3);
		      })) ? (e2 = r3[i3], o.setProperty(e2, n3, t3)) : ((r3 = {})[i3] = o.setProperty({}, n3, t3), s2.push(r3)) : ((e2 = {})[i3] = t3, s2.push(e2)), this;
		    }, a.prototype.setOptions = function(t3, i3) {
		      if (t3) for (var n3 = Object.keys(t3), r3 = 0; r3 < n3.length; ++r3) this.setOption(n3[r3], t3[n3[r3]], i3);
		      return this;
		    }, a.prototype.toString = function() {
		      var t3 = this.constructor.className, i3 = this.fullName;
		      return i3.length ? t3 + " " + i3 : t3;
		    }, a.prototype.c = function() {
		      return this.f && "proto3" !== this.f ? this.f : it;
		    }, a.p = function(t3) {
		      e = t3;
		    };
		  }, { 25: 25, 37: 37 }], 25: [function(t2, i2, n2) {
		    i2.exports = o;
		    var e = t2(24), r2 = (((o.prototype = Object.create(e.prototype)).constructor = o).className = "OneOf", t2(16)), s = t2(37);
		    function o(t3, i3, n3, r3) {
		      if (Array.isArray(i3) || (n3 = i3, i3 = it), e.call(this, t3, n3), i3 !== it && !Array.isArray(i3)) throw TypeError("fieldNames must be an Array");
		      this.oneof = i3 || [], this.fieldsArray = [], this.comment = r3;
		    }
		    function u(t3) {
		      if (t3.parent) for (var i3 = 0; i3 < t3.fieldsArray.length; ++i3) t3.fieldsArray[i3].parent || t3.parent.add(t3.fieldsArray[i3]);
		    }
		    o.fromJSON = function(t3, i3) {
		      return new o(t3, i3.oneof, i3.options, i3.comment);
		    }, o.prototype.toJSON = function(t3) {
		      t3 = !!t3 && !!t3.keepComments;
		      return s.toObject(["options", this.options, "oneof", this.oneof, "comment", t3 ? this.comment : it]);
		    }, o.prototype.add = function(t3) {
		      if (t3 instanceof r2) return t3.parent && t3.parent !== this.parent && t3.parent.remove(t3), this.oneof.push(t3.name), this.fieldsArray.push(t3), u(t3.partOf = this), this;
		      throw TypeError("field must be a Field");
		    }, o.prototype.remove = function(t3) {
		      if (!(t3 instanceof r2)) throw TypeError("field must be a Field");
		      var i3 = this.fieldsArray.indexOf(t3);
		      if (i3 < 0) throw Error(t3 + " is not a member of " + this);
		      return this.fieldsArray.splice(i3, 1), -1 < (i3 = this.oneof.indexOf(t3.name)) && this.oneof.splice(i3, 1), t3.partOf = null, this;
		    }, o.prototype.onAdd = function(t3) {
		      e.prototype.onAdd.call(this, t3);
		      for (var i3 = 0; i3 < this.oneof.length; ++i3) {
		        var n3 = t3.get(this.oneof[i3]);
		        n3 && !n3.partOf && (n3.partOf = this).fieldsArray.push(n3);
		      }
		      u(this);
		    }, o.prototype.onRemove = function(t3) {
		      for (var i3, n3 = 0; n3 < this.fieldsArray.length; ++n3) (i3 = this.fieldsArray[n3]).parent && i3.parent.remove(i3);
		      e.prototype.onRemove.call(this, t3);
		    }, Object.defineProperty(o.prototype, "isProto3Optional", { get: function() {
		      var t3;
		      return null != this.fieldsArray && 1 === this.fieldsArray.length && (null != (t3 = this.fieldsArray[0]).options && true === t3.options.proto3_optional);
		    } }), o.d = function() {
		      for (var n3 = Array(arguments.length), t3 = 0; t3 < arguments.length; ) n3[t3] = arguments[t3++];
		      return function(t4, i3) {
		        s.decorateType(t4.constructor).add(new o(i3, n3)), Object.defineProperty(t4, i3, { get: s.oneOfGetter(n3), set: s.oneOfSetter(n3) });
		      };
		    };
		  }, { 16: 16, 24: 24, 37: 37 }], 26: [function(t2, i2, n2) {
		    (i2.exports = tt).filename = null, tt.defaults = { keepCase: false };
		    var P = t2(34), V = t2(29), M = t2(35), R = t2(16), F = t2(20), $ = t2(25), D = t2(15), U = t2(33), C = t2(22), q = t2(24), X = t2(36), Y = t2(37), G = /^[1-9][0-9]*$/, z = /^-?[1-9][0-9]*$/, Z = /^0[x][0-9a-fA-F]+$/, B = /^-?0[x][0-9a-fA-F]+$/, H = /^0[0-7]+$/, K = /^-?0[0-7]+$/, W = /^(?![eE])[0-9]*(?:\.[0-9]*)?(?:[eE][+-]?[0-9]+)?$/, J = /^[a-zA-Z_][a-zA-Z_0-9]*$/, Q = /^(?:\.?[a-zA-Z_][a-zA-Z_0-9]*)(?:\.[a-zA-Z_][a-zA-Z_0-9]*)*$/;
		    function tt(t3, i3, n3) {
		      i3 instanceof V || (n3 = i3, i3 = new V());
		      var r2, e, s, h, o = (n3 = n3 || tt.defaults).preferTrailingComment || false, u = P(t3, n3.alternateCommentMode || false), c = u.next, f = u.push, l = u.peek, d = u.skip, a = u.cmnt, p = true, v = "proto2", b = i3, w = [], y = {}, m = n3.keepCase ? function(t4) {
		        return t4;
		      } : Y.camelCase;
		      function g(t4, i4, n4) {
		        var r3 = tt.filename;
		        return n4 || (tt.filename = null), Error("illegal " + (i4 || "token") + " '" + t4 + "' (" + (r3 ? r3 + ", " : "") + "line " + u.line + ")");
		      }
		      function j() {
		        var t4, i4 = [];
		        do {
		          if ('"' !== (t4 = c()) && "'" !== t4) throw g(t4);
		        } while (i4.push(c()), d(t4), '"' === (t4 = l()) || "'" === t4);
		        return i4.join("");
		      }
		      function E(i4) {
		        var n4 = c();
		        switch (n4) {
		          case "'":
		          case '"':
		            return f(n4), j();
		          case "true":
		          case "TRUE":
		            return true;
		          case "false":
		          case "FALSE":
		            return false;
		        }
		        try {
		          var t4 = n4, r3 = true, e2 = 1;
		          switch ("-" == (t4[0] || "") && (e2 = -1, t4 = t4.substring(1)), t4) {
		            case "inf":
		            case "INF":
		            case "Inf":
		              return e2 * (1 / 0);
		            case "nan":
		            case "NAN":
		            case "Nan":
		            case "NaN":
		              return NaN;
		            case "0":
		              return 0;
		          }
		          if (G.test(t4)) return e2 * parseInt(t4, 10);
		          if (Z.test(t4)) return e2 * parseInt(t4, 16);
		          if (H.test(t4)) return e2 * parseInt(t4, 8);
		          if (W.test(t4)) return e2 * parseFloat(t4);
		          throw g(t4, "number", r3);
		        } catch (t5) {
		          if (Q.test(n4)) return n4;
		          throw g(n4, "value");
		        }
		      }
		      function k(i4, n4) {
		        var t4;
		        do {
		          if (!n4 || '"' !== (r3 = l()) && "'" !== r3) try {
		            i4.push([t4 = O(c()), d("to", true) ? O(c()) : t4]);
		          } catch (t5) {
		            if (!(n4 && Q.test(r3) && 2023 <= v)) throw t5;
		            i4.push(r3);
		          }
		          else {
		            var r3 = j();
		            if (i4.push(r3), 2023 <= v) throw g(r3, "id");
		          }
		        } while (d(",", true));
		        var e2 = { options: it, setOption: function(t5, i5) {
		          this.options === it && (this.options = {}), this.options[t5] = i5;
		        } };
		        A(e2, function(t5) {
		          if ("option" !== t5) throw g(t5);
		          S(e2, t5), d(";");
		        }, function() {
		          N(e2);
		        });
		      }
		      function O(t4, i4) {
		        switch (t4) {
		          case "max":
		          case "MAX":
		          case "Max":
		            return 536870911;
		          case "0":
		            return 0;
		        }
		        if (i4 || "-" != (t4[0] || "")) {
		          if (z.test(t4)) return parseInt(t4, 10);
		          if (B.test(t4)) return parseInt(t4, 16);
		          if (K.test(t4)) return parseInt(t4, 8);
		        }
		        throw g(t4, "id");
		      }
		      function _(t4, i4) {
		        switch (i4) {
		          case "option":
		            return S(t4, i4), d(";"), true;
		          case "message":
		            return T(t4), true;
		          case "enum":
		            return L(t4), true;
		          case "export":
		          case "local":
		            return v < "2024" ? false : "export" !== (i4 = c()) && "local" !== i4 && (("message" === i4 || "enum" === i4) && _(t4, i4));
		          case "service":
		            var h2, n4 = t4, r3 = i4;
		            if (J.test(r3 = c())) return A(h2 = new U(r3), function(t5) {
		              if (!_(h2, t5)) {
		                if ("rpc" !== t5) throw g(t5);
		                var i5 = h2, n5 = a(), r4 = t5;
		                if (!J.test(t5 = c())) throw g(t5, "name");
		                var e3, s3, o2, u2 = t5;
		                if (d("("), d("stream", true) && (s3 = true), !Q.test(t5 = c())) throw g(t5);
		                if (e3 = t5, d(")"), d("returns"), d("("), d("stream", true) && (o2 = true), !Q.test(t5 = c())) throw g(t5);
		                t5 = t5, d(")");
		                var f2 = new C(u2, r4, e3, t5, s3, o2);
		                f2.comment = n5, A(f2, function(t6) {
		                  if ("option" !== t6) throw g(t6);
		                  S(f2, t6), d(";");
		                }), i5.add(f2);
		              }
		            }), n4.add(h2), n4 === b && w.push(h2), true;
		            throw g(r3, "service name");
		          case "extend":
		            var e2, s2 = t4, n4 = i4;
		            if (Q.test(n4 = c())) return e2 = n4, A(null, function(t5) {
		              switch (t5) {
		                case "required":
		                case "repeated":
		                  x(s2, t5, e2);
		                  break;
		                case "optional":
		                  x(s2, "proto3" === v ? "proto3_optional" : "optional", e2);
		                  break;
		                default:
		                  if ("proto2" === v || !Q.test(t5)) throw g(t5);
		                  f(t5), x(s2, "optional", e2);
		              }
		            }), true;
		            throw g(n4, "reference");
		        }
		        return false;
		      }
		      function A(t4, i4, n4) {
		        var r3, e2 = u.line;
		        if (t4 && ("string" != typeof t4.comment && (t4.comment = a()), t4.filename = tt.filename), d("{", true)) {
		          for (; "}" !== (r3 = c()); ) i4(r3);
		          d(";", true);
		        } else n4 && n4(), d(";"), t4 && ("string" != typeof t4.comment || o) && (t4.comment = a(e2) || t4.comment);
		      }
		      function T(t4, i4) {
		        if (!J.test(i4 = c())) throw g(i4, "type name");
		        var u2 = new M(i4);
		        A(u2, function(t5) {
		          if (!_(u2, t5)) switch (t5) {
		            case "map":
		              var i5 = u2, n4 = (d("<"), c());
		              if (X.mapKey[n4] === it) throw g(n4, "type");
		              d(",");
		              var r3 = c();
		              if (!Q.test(r3)) throw g(r3, "type");
		              d(">");
		              var e2 = c();
		              if (!J.test(e2)) throw g(e2, "name");
		              d("=");
		              var s2 = new F(m(e2), O(c()), n4, r3);
		              A(s2, function(t6) {
		                if ("option" !== t6) throw g(t6);
		                S(s2, t6), d(";");
		              }, function() {
		                N(s2);
		              }), i5.add(s2);
		              break;
		            case "required":
		              if ("proto2" !== v) throw g(t5);
		            case "repeated":
		              x(u2, t5);
		              break;
		            case "optional":
		              if ("proto3" === v) x(u2, "proto3_optional");
		              else {
		                if ("proto2" !== v) throw g(t5);
		                x(u2, "optional");
		              }
		              break;
		            case "oneof":
		              e2 = u2, n4 = t5;
		              if (!J.test(n4 = c())) throw g(n4, "name");
		              var o2 = new $(m(n4));
		              A(o2, function(t6) {
		                "option" === t6 ? (S(o2, t6), d(";")) : (f(t6), x(o2, "optional"));
		              }), e2.add(o2);
		              break;
		            case "extensions":
		              k(u2.extensions || (u2.extensions = []));
		              break;
		            case "reserved":
		              k(u2.reserved || (u2.reserved = []), true);
		              break;
		            default:
		              if ("proto2" === v || !Q.test(t5)) throw g(t5);
		              f(t5), x(u2, "optional");
		          }
		        }), t4.add(u2), t4 === b && w.push(u2);
		      }
		      function x(t4, i4, n4) {
		        var r3 = c();
		        if ("group" === r3) {
		          var e2 = t4, s2 = i4;
		          if (2023 <= v) throw g("group");
		          var o2, u2, f2 = c();
		          if (J.test(f2)) return u2 = Y.lcFirst(f2), f2 === u2 && (f2 = Y.ucFirst(f2)), d("="), h2 = O(c()), (o2 = new M(f2)).group = true, (u2 = new R(u2, h2, f2, s2)).filename = tt.filename, A(o2, function(t5) {
		            switch (t5) {
		              case "option":
		                S(o2, t5), d(";");
		                break;
		              case "required":
		              case "repeated":
		                x(o2, t5);
		                break;
		              case "optional":
		                x(o2, "proto3" === v ? "proto3_optional" : "optional");
		                break;
		              case "message":
		                T(o2);
		                break;
		              case "enum":
		                L(o2);
		                break;
		              case "reserved":
		                k(o2.reserved || (o2.reserved = []), true);
		                break;
		              case "export":
		              case "local":
		                if (v < "2024") throw g(t5);
		                switch (t5 = c()) {
		                  case "message":
		                  case "enum":
		                    T(o2);
		                    break;
		                  default:
		                    throw g(t5);
		                }
		                break;
		              default:
		                throw g(t5);
		            }
		          }), void e2.add(o2).add(u2);
		          throw g(f2, "name");
		        }
		        for (; r3.endsWith(".") || l().startsWith("."); ) r3 += c();
		        if (!Q.test(r3)) throw g(r3, "type");
		        var h2 = c();
		        if (!J.test(h2)) throw g(h2, "name");
		        h2 = m(h2), d("=");
		        var a2 = new R(h2, O(c()), r3, i4, n4);
		        A(a2, function(t5) {
		          if ("option" !== t5) throw g(t5);
		          S(a2, t5), d(";");
		        }, function() {
		          N(a2);
		        }), "proto3_optional" === i4 ? (s2 = new $("_" + h2), a2.setOption("proto3_optional", true), s2.add(a2), t4.add(s2)) : t4.add(a2), t4 === b && w.push(a2);
		      }
		      function L(t4, i4) {
		        if (!J.test(i4 = c())) throw g(i4, "name");
		        var s2 = new D(i4);
		        A(s2, function(t5) {
		          switch (t5) {
		            case "option":
		              S(s2, t5), d(";");
		              break;
		            case "reserved":
		              k(s2.reserved || (s2.reserved = []), true), s2.reserved === it && (s2.reserved = []);
		              break;
		            default:
		              var i5 = s2, n4 = t5;
		              if (!J.test(n4)) throw g(n4, "name");
		              d("=");
		              var r3 = O(c(), true), e2 = { options: it, getOption: function(t6) {
		                return this.options[t6];
		              }, setOption: function(t6, i6) {
		                q.prototype.setOption.call(e2, t6, i6);
		              }, setParsedOption: function() {
		                return it;
		              } };
		              return A(e2, function(t6) {
		                if ("option" !== t6) throw g(t6);
		                S(e2, t6), d(";");
		              }, function() {
		                N(e2);
		              }), void i5.add(n4, r3, e2.comment, e2.parsedOptions || e2.options);
		          }
		        }), t4.add(s2), t4 === b && w.push(s2);
		      }
		      function S(t4, i4) {
		        var n4 = true;
		        for ("option" === i4 && (i4 = c()); "=" !== i4; ) {
		          if ("(" === i4 && (r3 = c(), d(")"), i4 = "(" + r3 + ")"), n4) {
		            if (n4 = false, i4.includes(".") && !i4.includes("(")) {
		              var r3 = i4.split("."), e2 = r3[0] + ".";
		              i4 = r3[1];
		              continue;
		            }
		            e2 = i4;
		          } else f2 = f2 ? f2 + i4 : i4;
		          i4 = c();
		        }
		        var s2, o2, u2 = f2 ? e2.concat(f2) : e2, u2 = (function t5(i5, n5) {
		          if (d("{", true)) {
		            for (var r4 = {}; !d("}", true); ) {
		              if (!J.test(h = c())) throw g(h, "name");
		              if (null === h) throw g(h, "end of input");
		              var e3, s3, o3 = h;
		              if (d(":", true), "{" === l()) e3 = t5(i5, n5 + "." + h);
		              else if ("[" === l()) {
		                if (e3 = [], d("[", true)) {
		                  for (; s3 = E(), e3.push(s3), d(",", true); ) ;
		                  d("]"), void 0 !== s3 && I(i5, n5 + "." + h, s3);
		                }
		              } else e3 = E(), I(i5, n5 + "." + h, e3);
		              var u3 = r4[o3];
		              u3 && (e3 = [].concat(u3).concat(e3)), r4[o3] = e3, d(",", true), d(";", true);
		            }
		            return r4;
		          }
		          var f3 = E();
		          I(i5, n5, f3);
		          return f3;
		        })(t4, u2), f2 = f2 && "." === f2[0] ? f2.slice(1) : f2;
		        e2 = e2 && "." === e2[e2.length - 1] ? e2.slice(0, -1) : e2, s2 = e2, u2 = u2, o2 = f2, (t4 = t4).setParsedOption && t4.setParsedOption(s2, u2, o2);
		      }
		      function I(t4, i4, n4) {
		        b === t4 && /^features\./.test(i4) ? y[i4] = n4 : t4.setOption && t4.setOption(i4, n4);
		      }
		      function N(t4) {
		        if (d("[", true)) {
		          for (; S(t4, "option"), d(",", true); ) ;
		          d("]");
		        }
		      }
		      for (; null !== (h = c()); ) switch (h) {
		        case "package":
		          if (!p) throw g(h);
		          if (r2 !== it) throw g("package");
		          if (r2 = c(), !Q.test(r2)) throw g(r2, "name");
		          b = b.define(r2), d(";");
		          break;
		        case "import":
		          if (!p) throw g(h);
		          !(function() {
		            var t4, i4;
		            switch (l()) {
		              case "option":
		                if (v < "2024") throw g("option");
		                return c(), j(), d(";");
		              case "weak":
		                i4 = s = s || [], c();
		                break;
		              case "public":
		                c();
		              default:
		                i4 = e = e || [];
		            }
		            t4 = j(), d(";"), i4.push(t4);
		          })();
		          break;
		        case "syntax":
		          if (!p) throw g(h);
		          if (d("="), (v = j()) < 2023) throw g(v, "syntax");
		          d(";");
		          break;
		        case "edition":
		          if (!p) throw g(h);
		          if (d("="), v = j(), !["2023", "2024"].includes(v)) throw g(v, "edition");
		          d(";");
		          break;
		        case "option":
		          S(b, h), d(";", true);
		          break;
		        default:
		          if (_(b, h)) {
		            p = false;
		            continue;
		          }
		          throw g(h);
		      }
		      return w.forEach((i4) => {
		        i4.f = v, Object.keys(y).forEach((t4) => {
		          i4.getOption(t4) === it && i4.setOption(t4, y[t4], true);
		        });
		      }), tt.filename = null, { package: r2, imports: e, weakImports: s, root: i3 };
		    }
		  }, { 15: 15, 16: 16, 20: 20, 22: 22, 24: 24, 25: 25, 29: 29, 33: 33, 34: 34, 35: 35, 36: 36, 37: 37 }], 27: [function(t2, i2, n2) {
		    i2.exports = f;
		    var r2, e = t2(39), s = e.LongBits, o = e.utf8;
		    function u(t3, i3) {
		      return RangeError("index out of range: " + t3.pos + " + " + (i3 || 1) + " > " + t3.len);
		    }
		    function f(t3) {
		      this.buf = t3, this.pos = 0, this.len = t3.length;
		    }
		    function h() {
		      return e.Buffer ? function(t3) {
		        return (f.create = function(t4) {
		          return e.Buffer.isBuffer(t4) ? new r2(t4) : c(t4);
		        })(t3);
		      } : c;
		    }
		    var a, c = "undefined" != typeof Uint8Array ? function(t3) {
		      if (t3 instanceof Uint8Array || Array.isArray(t3)) return new f(t3);
		      throw Error("illegal buffer");
		    } : function(t3) {
		      if (Array.isArray(t3)) return new f(t3);
		      throw Error("illegal buffer");
		    };
		    function l() {
		      var t3 = new s(0, 0), i3 = 0;
		      if (!(4 < this.len - this.pos)) {
		        for (; i3 < 3; ++i3) {
		          if (this.pos >= this.len) throw u(this);
		          if (t3.lo = (t3.lo | (127 & this.buf[this.pos]) << 7 * i3) >>> 0, this.buf[this.pos++] < 128) return t3;
		        }
		        return t3.lo = (t3.lo | (127 & this.buf[this.pos++]) << 7 * i3) >>> 0, t3;
		      }
		      for (; i3 < 4; ++i3) if (t3.lo = (t3.lo | (127 & this.buf[this.pos]) << 7 * i3) >>> 0, this.buf[this.pos++] < 128) return t3;
		      if (t3.lo = (t3.lo | (127 & this.buf[this.pos]) << 28) >>> 0, t3.hi = (t3.hi | (127 & this.buf[this.pos]) >> 4) >>> 0, this.buf[this.pos++] < 128) return t3;
		      if (i3 = 0, 4 < this.len - this.pos) {
		        for (; i3 < 5; ++i3) if (t3.hi = (t3.hi | (127 & this.buf[this.pos]) << 7 * i3 + 3) >>> 0, this.buf[this.pos++] < 128) return t3;
		      } else for (; i3 < 5; ++i3) {
		        if (this.pos >= this.len) throw u(this);
		        if (t3.hi = (t3.hi | (127 & this.buf[this.pos]) << 7 * i3 + 3) >>> 0, this.buf[this.pos++] < 128) return t3;
		      }
		      throw Error("invalid varint encoding");
		    }
		    function d(t3, i3) {
		      return (t3[i3 - 4] | t3[i3 - 3] << 8 | t3[i3 - 2] << 16 | t3[i3 - 1] << 24) >>> 0;
		    }
		    function p() {
		      if (this.pos + 8 > this.len) throw u(this, 8);
		      return new s(d(this.buf, this.pos += 4), d(this.buf, this.pos += 4));
		    }
		    f.create = h(), f.prototype.T = e.Array.prototype.subarray || e.Array.prototype.slice, f.prototype.uint32 = (a = 4294967295, function() {
		      if (a = (127 & this.buf[this.pos]) >>> 0, this.buf[this.pos++] < 128 || (a = (a | (127 & this.buf[this.pos]) << 7) >>> 0, this.buf[this.pos++] < 128 || (a = (a | (127 & this.buf[this.pos]) << 14) >>> 0, this.buf[this.pos++] < 128 || (a = (a | (127 & this.buf[this.pos]) << 21) >>> 0, this.buf[this.pos++] < 128 || (a = (a | (15 & this.buf[this.pos]) << 28) >>> 0, this.buf[this.pos++] < 128 || !((this.pos += 5) > this.len)))))) return a;
		      throw this.pos = this.len, u(this, 10);
		    }), f.prototype.int32 = function() {
		      return 0 | this.uint32();
		    }, f.prototype.sint32 = function() {
		      var t3 = this.uint32();
		      return t3 >>> 1 ^ -(1 & t3) | 0;
		    }, f.prototype.bool = function() {
		      return 0 !== this.uint32();
		    }, f.prototype.fixed32 = function() {
		      if (this.pos + 4 > this.len) throw u(this, 4);
		      return d(this.buf, this.pos += 4);
		    }, f.prototype.sfixed32 = function() {
		      if (this.pos + 4 > this.len) throw u(this, 4);
		      return 0 | d(this.buf, this.pos += 4);
		    }, f.prototype.float = function() {
		      if (this.pos + 4 > this.len) throw u(this, 4);
		      var t3 = e.float.readFloatLE(this.buf, this.pos);
		      return this.pos += 4, t3;
		    }, f.prototype.double = function() {
		      if (this.pos + 8 > this.len) throw u(this, 4);
		      var t3 = e.float.readDoubleLE(this.buf, this.pos);
		      return this.pos += 8, t3;
		    }, f.prototype.bytes = function() {
		      var t3 = this.uint32(), i3 = this.pos, n3 = this.pos + t3;
		      if (n3 > this.len) throw u(this, t3);
		      return this.pos += t3, Array.isArray(this.buf) ? this.buf.slice(i3, n3) : i3 === n3 ? (t3 = e.Buffer) ? t3.alloc(0) : new this.buf.constructor(0) : this.T.call(this.buf, i3, n3);
		    }, f.prototype.string = function() {
		      var t3 = this.bytes();
		      return o.read(t3, 0, t3.length);
		    }, f.prototype.skip = function(t3) {
		      if ("number" == typeof t3) {
		        if (this.pos + t3 > this.len) throw u(this, t3);
		        this.pos += t3;
		      } else do {
		        if (this.pos >= this.len) throw u(this);
		      } while (128 & this.buf[this.pos++]);
		      return this;
		    }, f.prototype.skipType = function(t3) {
		      switch (t3) {
		        case 0:
		          this.skip();
		          break;
		        case 1:
		          this.skip(8);
		          break;
		        case 2:
		          this.skip(this.uint32());
		          break;
		        case 3:
		          for (; 4 != (t3 = 7 & this.uint32()); ) this.skipType(t3);
		          break;
		        case 5:
		          this.skip(4);
		          break;
		        default:
		          throw Error("invalid wire type " + t3 + " at offset " + this.pos);
		      }
		      return this;
		    }, f.p = function(t3) {
		      r2 = t3, f.create = h(), r2.p();
		      var i3 = e.Long ? "toLong" : "toNumber";
		      e.merge(f.prototype, { int64: function() {
		        return l.call(this)[i3](false);
		      }, uint64: function() {
		        return l.call(this)[i3](true);
		      }, sint64: function() {
		        return l.call(this).zzDecode()[i3](false);
		      }, fixed64: function() {
		        return p.call(this)[i3](true);
		      }, sfixed64: function() {
		        return p.call(this)[i3](false);
		      } });
		    };
		  }, { 39: 39 }], 28: [function(t2, i2, n2) {
		    i2.exports = s;
		    var r2 = t2(27), e = ((s.prototype = Object.create(r2.prototype)).constructor = s, t2(39));
		    function s(t3) {
		      r2.call(this, t3);
		    }
		    s.p = function() {
		      e.Buffer && (s.prototype.T = e.Buffer.prototype.slice);
		    }, s.prototype.string = function() {
		      var t3 = this.uint32();
		      return this.buf.utf8Slice ? this.buf.utf8Slice(this.pos, this.pos = Math.min(this.pos + t3, this.len)) : this.buf.toString("utf-8", this.pos, this.pos = Math.min(this.pos + t3, this.len));
		    }, s.p();
		  }, { 27: 27, 39: 39 }], 29: [function(t2, i2, n2) {
		    i2.exports = f;
		    var r2, d, p, e = t2(23), s = (((f.prototype = Object.create(e.prototype)).constructor = f).className = "Root", t2(16)), o = t2(15), u = t2(25), v = t2(37);
		    function f(t3) {
		      e.call(this, "", t3), this.deferred = [], this.files = [], this.f = "proto2", this.j = {};
		    }
		    function b() {
		    }
		    f.fromJSON = function(t3, i3) {
		      return i3 = i3 || new f(), t3.options && i3.setOptions(t3.options), i3.addJSON(t3.nested).resolveAll();
		    }, f.prototype.resolvePath = v.path.resolve, f.prototype.fetch = v.fetch, f.prototype.load = function t3(i3, s2, e2) {
		      "function" == typeof s2 && (e2 = s2, s2 = it);
		      var o2 = this;
		      if (!e2) return v.asPromise(t3, o2, i3, s2);
		      var u2 = e2 === b;
		      function f2(t4, i4) {
		        if (e2) {
		          if (u2) throw t4;
		          i4 && i4.resolveAll();
		          var n4 = e2;
		          e2 = null, n4(t4, i4);
		        }
		      }
		      function h2(t4) {
		        var i4 = t4.lastIndexOf("google/protobuf/");
		        if (-1 < i4) {
		          t4 = t4.substring(i4);
		          if (t4 in p) return t4;
		        }
		        return null;
		      }
		      function a2(t4, i4) {
		        try {
		          if (v.isString(i4) && "{" == (i4[0] || "") && (i4 = JSON.parse(i4)), v.isString(i4)) {
		            d.filename = t4;
		            var n4, r4 = d(i4, o2, s2), e3 = 0;
		            if (r4.imports) for (; e3 < r4.imports.length; ++e3) (n4 = h2(r4.imports[e3]) || o2.resolvePath(t4, r4.imports[e3])) && c(n4);
		            if (r4.weakImports) for (e3 = 0; e3 < r4.weakImports.length; ++e3) (n4 = h2(r4.weakImports[e3]) || o2.resolvePath(t4, r4.weakImports[e3])) && c(n4, true);
		          } else o2.setOptions(i4.options).addJSON(i4.nested);
		        } catch (t5) {
		          f2(t5);
		        }
		        u2 || l || f2(null, o2);
		      }
		      function c(n4, r4) {
		        if (n4 = h2(n4) || n4, !~o2.files.indexOf(n4)) if (o2.files.push(n4), n4 in p) u2 ? a2(n4, p[n4]) : (++l, setTimeout(function() {
		          --l, a2(n4, p[n4]);
		        }));
		        else if (u2) {
		          var t4;
		          try {
		            t4 = v.fs.readFileSync(n4).toString("utf8");
		          } catch (t5) {
		            return void (r4 || f2(t5));
		          }
		          a2(n4, t4);
		        } else ++l, o2.fetch(n4, function(t5, i4) {
		          --l, e2 && (t5 ? r4 ? l || f2(null, o2) : f2(t5) : a2(n4, i4));
		        });
		      }
		      var l = 0;
		      v.isString(i3) && (i3 = [i3]);
		      for (var n3, r3 = 0; r3 < i3.length; ++r3) (n3 = o2.resolvePath("", i3[r3])) && c(n3);
		      return u2 ? o2.resolveAll() : l || f2(null, o2), o2;
		    }, f.prototype.loadSync = function(t3, i3) {
		      if (v.isNode) return this.load(t3, i3, b);
		      throw Error("not supported");
		    }, f.prototype.resolveAll = function() {
		      if (!this.y) return this;
		      if (this.deferred.length) throw Error("unresolvable extensions: " + this.deferred.map(function(t3) {
		        return "'extend " + t3.extend + "' in " + t3.parent.fullName;
		      }).join(", "));
		      return e.prototype.resolveAll.call(this);
		    };
		    var h = /^[A-Z]/;
		    function a(t3, i3) {
		      var n3, r3 = i3.parent.lookup(i3.extend);
		      if (r3) return n3 = new s(i3.fullName, i3.id, i3.type, i3.rule, it, i3.options), r3.get(n3.name) || ((n3.declaringField = i3).extensionField = n3, r3.add(n3)), 1;
		    }
		    f.prototype._ = function(t3) {
		      if (t3 instanceof s) t3.extend === it || t3.extensionField || a(0, t3) || this.deferred.push(t3);
		      else if (t3 instanceof o) h.test(t3.name) && (t3.parent[t3.name] = t3.values);
		      else if (!(t3 instanceof u)) {
		        if (t3 instanceof r2) for (var i3 = 0; i3 < this.deferred.length; ) a(0, this.deferred[i3]) ? this.deferred.splice(i3, 1) : ++i3;
		        for (var n3 = 0; n3 < t3.nestedArray.length; ++n3) this._(t3.v[n3]);
		        h.test(t3.name) && (t3.parent[t3.name] = t3);
		      }
		      (t3 instanceof r2 || t3 instanceof o || t3 instanceof s) && (this.j[t3.fullName] = t3);
		    }, f.prototype.A = function(t3) {
		      var i3;
		      if (t3 instanceof s) t3.extend !== it && (t3.extensionField ? (t3.extensionField.parent.remove(t3.extensionField), t3.extensionField = null) : -1 < (i3 = this.deferred.indexOf(t3)) && this.deferred.splice(i3, 1));
		      else if (t3 instanceof o) h.test(t3.name) && delete t3.parent[t3.name];
		      else if (t3 instanceof e) {
		        for (var n3 = 0; n3 < t3.nestedArray.length; ++n3) this.A(t3.v[n3]);
		        h.test(t3.name) && delete t3.parent[t3.name];
		      }
		      delete this.j[t3.fullName];
		    }, f.p = function(t3, i3, n3) {
		      r2 = t3, d = i3, p = n3;
		    };
		  }, { 15: 15, 16: 16, 23: 23, 25: 25, 37: 37 }], 30: [function(t2, i2, n2) {
		    i2.exports = {};
		  }, {}], 31: [function(t2, i2, n2) {
		    n2.Service = t2(32);
		  }, { 32: 32 }], 32: [function(t2, i2, n2) {
		    i2.exports = r2;
		    var u = t2(39);
		    function r2(t3, i3, n3) {
		      if ("function" != typeof t3) throw TypeError("rpcImpl must be a function");
		      u.EventEmitter.call(this), this.rpcImpl = t3, this.requestDelimited = !!i3, this.responseDelimited = !!n3;
		    }
		    ((r2.prototype = Object.create(u.EventEmitter.prototype)).constructor = r2).prototype.rpcCall = function t3(n3, i3, r3, e, s) {
		      if (!e) throw TypeError("request must be specified");
		      var o = this;
		      if (!s) return u.asPromise(t3, o, n3, i3, r3, e);
		      if (!o.rpcImpl) return setTimeout(function() {
		        s(Error("already ended"));
		      }, 0), it;
		      try {
		        return o.rpcImpl(n3, i3[o.requestDelimited ? "encodeDelimited" : "encode"](e).finish(), function(t4, i4) {
		          if (t4) return o.emit("error", t4, n3), s(t4);
		          if (null === i4) return o.end(true), it;
		          if (!(i4 instanceof r3)) try {
		            i4 = r3[o.responseDelimited ? "decodeDelimited" : "decode"](i4);
		          } catch (t5) {
		            return o.emit("error", t5, n3), s(t5);
		          }
		          return o.emit("data", i4, n3), s(null, i4);
		        });
		      } catch (t4) {
		        return o.emit("error", t4, n3), setTimeout(function() {
		          s(t4);
		        }, 0), it;
		      }
		    }, r2.prototype.end = function(t3) {
		      return this.rpcImpl && (t3 || this.rpcImpl(null, null, null), this.rpcImpl = null, this.emit("end").off()), this;
		    };
		  }, { 39: 39 }], 33: [function(t2, i2, n2) {
		    i2.exports = o;
		    var r2 = t2(23), s = (((o.prototype = Object.create(r2.prototype)).constructor = o).className = "Service", t2(22)), u = t2(37), f = t2(31);
		    function o(t3, i3) {
		      r2.call(this, t3, i3), this.methods = {}, this.x = null;
		    }
		    function e(t3) {
		      return t3.x = null, t3;
		    }
		    o.fromJSON = function(t3, i3) {
		      var n3 = new o(t3, i3.options);
		      if (i3.methods) for (var r3 = Object.keys(i3.methods), e2 = 0; e2 < r3.length; ++e2) n3.add(s.fromJSON(r3[e2], i3.methods[r3[e2]]));
		      return i3.nested && n3.addJSON(i3.nested), i3.edition && (n3.f = i3.edition), n3.comment = i3.comment, n3.a = "proto3", n3;
		    }, o.prototype.toJSON = function(t3) {
		      var i3 = r2.prototype.toJSON.call(this, t3), n3 = !!t3 && !!t3.keepComments;
		      return u.toObject(["edition", this.c(), "options", i3 && i3.options || it, "methods", r2.arrayToJSON(this.methodsArray, t3) || {}, "nested", i3 && i3.nested || it, "comment", n3 ? this.comment : it]);
		    }, Object.defineProperty(o.prototype, "methodsArray", { get: function() {
		      return this.x || (this.x = u.toArray(this.methods));
		    } }), o.prototype.get = function(t3) {
		      return this.methods[t3] || r2.prototype.get.call(this, t3);
		    }, o.prototype.resolveAll = function() {
		      if (this.y) {
		        r2.prototype.resolve.call(this);
		        for (var t3 = this.methodsArray, i3 = 0; i3 < t3.length; ++i3) t3[i3].resolve();
		      }
		      return this;
		    }, o.prototype.g = function(i3) {
		      return this.w && (i3 = this.f || i3, r2.prototype.g.call(this, i3), this.methodsArray.forEach((t3) => {
		        t3.g(i3);
		      })), this;
		    }, o.prototype.add = function(t3) {
		      if (this.get(t3.name)) throw Error("duplicate name '" + t3.name + "' in " + this);
		      return t3 instanceof s ? e((this.methods[t3.name] = t3).parent = this) : r2.prototype.add.call(this, t3);
		    }, o.prototype.remove = function(t3) {
		      if (t3 instanceof s) {
		        if (this.methods[t3.name] !== t3) throw Error(t3 + " is not a member of " + this);
		        return delete this.methods[t3.name], t3.parent = null, e(this);
		      }
		      return r2.prototype.remove.call(this, t3);
		    }, o.prototype.create = function(t3, i3, n3) {
		      for (var r3, e2 = new f.Service(t3, i3, n3), s2 = 0; s2 < this.methodsArray.length; ++s2) {
		        var o2 = u.lcFirst((r3 = this.x[s2]).resolve().name).replace(/[^$\w_]/g, "");
		        e2[o2] = u.codegen(["r", "c"], u.isReserved(o2) ? o2 + "_" : o2)("return this.rpcCall(m,q,s,r,c)")({ m: r3, q: r3.resolvedRequestType.ctor, s: r3.resolvedResponseType.ctor });
		      }
		      return e2;
		    };
		  }, { 22: 22, 23: 23, 31: 31, 37: 37 }], 34: [function(t2, i2, n2) {
		    i2.exports = s;
		    var k = /[\s{}=;:[\],'"()<>]/g, O = /(?:"([^"\\]*(?:\\.[^"\\]*)*)")/g, _ = /(?:'([^'\\]*(?:\\.[^'\\]*)*)')/g, A = /^ *[*/]+ */, T = /^\s*\*?\/*/, x = /\n/g, L = /\s/, r2 = /\\(.?)/g, e = { 0: "\0", r: "\r", n: "\n", t: "	" };
		    function S(t3) {
		      return t3.replace(r2, function(t4, i3) {
		        switch (i3) {
		          case "\\":
		          case "":
		            return i3;
		          default:
		            return e[i3] || "";
		        }
		      });
		    }
		    function s(h, a) {
		      h = h.toString();
		      var c = 0, l = h.length, d = 1, f = 0, p = {}, v = [], b = null;
		      function w(t3) {
		        return Error("illegal " + t3 + " (line " + d + ")");
		      }
		      function y(t3) {
		        return h[0 | t3] || "";
		      }
		      function m(t3, i3, n3) {
		        var r4, e3 = { type: h[0 | t3++] || "", lineEmpty: false, leading: n3 }, n3 = a ? 2 : 3, s2 = t3 - n3;
		        do {
		          if (--s2 < 0 || "\n" == (r4 = h[0 | s2] || "")) {
		            e3.lineEmpty = true;
		            break;
		          }
		        } while (" " === r4 || "	" === r4);
		        for (var o = h.substring(t3, i3).split(x), u = 0; u < o.length; ++u) o[u] = o[u].replace(a ? T : A, "").trim();
		        e3.text = o.join("\n").trim(), p[d] = e3, f = d;
		      }
		      function g(t3) {
		        var i3 = j(t3), t3 = h.substring(t3, i3);
		        return /^\s*\/\//.test(t3);
		      }
		      function j(t3) {
		        for (var i3 = t3; i3 < l && "\n" !== y(i3); ) i3++;
		        return i3;
		      }
		      function r3() {
		        if (0 < v.length) return v.shift();
		        if (b) {
		          var t3 = "'" === b ? _ : O, i3 = (t3.lastIndex = c - 1, t3.exec(h));
		          if (i3) return c = t3.lastIndex, E(b), b = null, S(i3[1]);
		          throw w("string");
		        }
		        var n3, r4, e3, s2, o, u = 0 === c;
		        do {
		          if (c === l) return null;
		          for (n3 = false; L.test(e3 = y(c)); ) if ("\n" === e3 && (u = true, ++d), ++c === l) return null;
		          if ("/" === y(c)) {
		            if (++c === l) throw w("comment");
		            if ("/" === y(c)) if (a) {
		              if (o = false, g((s2 = c) - 1)) for (o = true; (c = j(c)) !== l && (c++, u && g(c)); ) ;
		              else c = Math.min(l, j(c) + 1);
		              o && (m(s2, c, u), u = true), d++;
		            } else {
		              for (o = "/" === y(s2 = c + 1); "\n" !== y(++c); ) if (c === l) return null;
		              ++c, o && (m(s2, c - 1, u), u = true), ++d;
		            }
		            else {
		              if ("*" !== (e3 = y(c))) return "/";
		              s2 = c + 1, o = a || "*" === y(s2);
		              do {
		                if ("\n" === e3 && ++d, ++c === l) throw w("comment");
		              } while (r4 = e3, e3 = y(c), "*" !== r4 || "/" !== e3);
		              ++c, o && (m(s2, c - 2, u), u = true);
		            }
		            n3 = true;
		          }
		        } while (n3);
		        var f2 = c;
		        if (k.lastIndex = 0, !k.test(y(f2++))) for (; f2 < l && !k.test(y(f2)); ) ++f2;
		        t3 = h.substring(c, c = f2);
		        return '"' != t3 && "'" != t3 || (b = t3), t3;
		      }
		      function E(t3) {
		        v.push(t3);
		      }
		      function e2() {
		        if (!v.length) {
		          var t3 = r3();
		          if (null === t3) return null;
		          E(t3);
		        }
		        return v[0];
		      }
		      return Object.defineProperty({ next: r3, peek: e2, push: E, skip: function(t3, i3) {
		        var n3 = e2();
		        if (n3 === t3) return r3(), true;
		        if (i3) return false;
		        throw w("token '" + n3 + "', '" + t3 + "' expected");
		      }, cmnt: function(t3) {
		        var i3, n3 = null;
		        return t3 === it ? (i3 = p[d - 1], delete p[d - 1], i3 && (a || "*" === i3.type || i3.lineEmpty) && (n3 = i3.leading ? i3.text : null)) : (f < t3 && e2(), i3 = p[t3], delete p[t3], !i3 || i3.lineEmpty || !a && "/" !== i3.type || (n3 = i3.leading ? null : i3.text)), n3;
		      } }, "line", { get: function() {
		        return d;
		      } });
		    }
		    s.unescape = S;
		  }, {}], 35: [function(t2, i2, n2) {
		    i2.exports = m;
		    var o = t2(23), u = (((m.prototype = Object.create(o.prototype)).constructor = m).className = "Type", t2(15)), f = t2(25), h = t2(16), a = t2(20), c = t2(33), e = t2(21), s = t2(27), l = t2(42), d = t2(37), p = t2(14), v = t2(13), b = t2(40), w = t2(12), y = t2(41);
		    function m(t3, i3) {
		      o.call(this, t3, i3), this.fields = {}, this.oneofs = it, this.extensions = it, this.reserved = it, this.group = it, this.L = null, this.e = null, this.S = null, this.I = null;
		    }
		    function r2(t3) {
		      return t3.L = t3.e = t3.S = null, delete t3.encode, delete t3.decode, delete t3.verify, t3;
		    }
		    Object.defineProperties(m.prototype, { fieldsById: { get: function() {
		      if (!this.L) {
		        this.L = {};
		        for (var t3 = Object.keys(this.fields), i3 = 0; i3 < t3.length; ++i3) {
		          var n3 = this.fields[t3[i3]], r3 = n3.id;
		          if (this.L[r3]) throw Error("duplicate id " + r3 + " in " + this);
		          this.L[r3] = n3;
		        }
		      }
		      return this.L;
		    } }, fieldsArray: { get: function() {
		      return this.e || (this.e = d.toArray(this.fields));
		    } }, oneofsArray: { get: function() {
		      return this.S || (this.S = d.toArray(this.oneofs));
		    } }, ctor: { get: function() {
		      return this.I || (this.ctor = m.generateConstructor(this)());
		    }, set: function(t3) {
		      for (var i3 = t3.prototype, n3 = (i3 instanceof e || ((t3.prototype = new e()).constructor = t3, d.merge(t3.prototype, i3)), t3.$type = t3.prototype.$type = this, d.merge(t3, e, true), this.I = t3, 0); n3 < this.fieldsArray.length; ++n3) this.e[n3].resolve();
		      for (var r3 = {}, n3 = 0; n3 < this.oneofsArray.length; ++n3) r3[this.S[n3].resolve().name] = { get: d.oneOfGetter(this.S[n3].oneof), set: d.oneOfSetter(this.S[n3].oneof) };
		      n3 && Object.defineProperties(t3.prototype, r3);
		    } } }), m.generateConstructor = function(t3) {
		      for (var i3, n3 = d.codegen(["p"], t3.name), r3 = 0; r3 < t3.fieldsArray.length; ++r3) (i3 = t3.e[r3]).map ? n3("this%s={}", d.safeProp(i3.name)) : i3.repeated && n3("this%s=[]", d.safeProp(i3.name));
		      return n3("if(p)for(var ks=Object.keys(p),i=0;i<ks.length;++i)if(p[ks[i]]!=null)")("this[ks[i]]=p[ks[i]]");
		    }, m.fromJSON = function(t3, i3) {
		      for (var n3 = new m(t3, i3.options), r3 = (n3.extensions = i3.extensions, n3.reserved = i3.reserved, Object.keys(i3.fields)), e2 = 0; e2 < r3.length; ++e2) n3.add((void 0 !== i3.fields[r3[e2]].keyType ? a : h).fromJSON(r3[e2], i3.fields[r3[e2]]));
		      if (i3.oneofs) for (r3 = Object.keys(i3.oneofs), e2 = 0; e2 < r3.length; ++e2) n3.add(f.fromJSON(r3[e2], i3.oneofs[r3[e2]]));
		      if (i3.nested) for (r3 = Object.keys(i3.nested), e2 = 0; e2 < r3.length; ++e2) {
		        var s2 = i3.nested[r3[e2]];
		        n3.add((s2.id !== it ? h : s2.fields !== it ? m : s2.values !== it ? u : s2.methods !== it ? c : o).fromJSON(r3[e2], s2));
		      }
		      return i3.extensions && i3.extensions.length && (n3.extensions = i3.extensions), i3.reserved && i3.reserved.length && (n3.reserved = i3.reserved), i3.group && (n3.group = true), i3.comment && (n3.comment = i3.comment), i3.edition && (n3.f = i3.edition), n3.a = "proto3", n3;
		    }, m.prototype.toJSON = function(t3) {
		      var i3 = o.prototype.toJSON.call(this, t3), n3 = !!t3 && !!t3.keepComments;
		      return d.toObject(["edition", this.c(), "options", i3 && i3.options || it, "oneofs", o.arrayToJSON(this.oneofsArray, t3), "fields", o.arrayToJSON(this.fieldsArray.filter(function(t4) {
		        return !t4.declaringField;
		      }), t3) || {}, "extensions", this.extensions && this.extensions.length ? this.extensions : it, "reserved", this.reserved && this.reserved.length ? this.reserved : it, "group", this.group || it, "nested", i3 && i3.nested || it, "comment", n3 ? this.comment : it]);
		    }, m.prototype.resolveAll = function() {
		      if (this.y) {
		        o.prototype.resolveAll.call(this);
		        for (var t3 = this.oneofsArray, i3 = 0; i3 < t3.length; ) t3[i3++].resolve();
		        for (var n3 = this.fieldsArray, i3 = 0; i3 < n3.length; ) n3[i3++].resolve();
		      }
		      return this;
		    }, m.prototype.g = function(i3) {
		      return this.w && (i3 = this.f || i3, o.prototype.g.call(this, i3), this.oneofsArray.forEach((t3) => {
		        t3.u(i3);
		      }), this.fieldsArray.forEach((t3) => {
		        t3.u(i3);
		      })), this;
		    }, m.prototype.get = function(t3) {
		      return this.fields[t3] || this.oneofs && this.oneofs[t3] || this.nested && this.nested[t3] || null;
		    }, m.prototype.add = function(t3) {
		      if (this.get(t3.name)) throw Error("duplicate name '" + t3.name + "' in " + this);
		      if (t3 instanceof h && t3.extend === it) {
		        if ((this.L || this.fieldsById)[t3.id]) throw Error("duplicate id " + t3.id + " in " + this);
		        if (this.isReservedId(t3.id)) throw Error("id " + t3.id + " is reserved in " + this);
		        if (this.isReservedName(t3.name)) throw Error("name '" + t3.name + "' is reserved in " + this);
		        return t3.parent && t3.parent.remove(t3), (this.fields[t3.name] = t3).message = this, t3.onAdd(this), r2(this);
		      }
		      return t3 instanceof f ? (this.oneofs || (this.oneofs = {}), (this.oneofs[t3.name] = t3).onAdd(this), r2(this)) : o.prototype.add.call(this, t3);
		    }, m.prototype.remove = function(t3) {
		      if (t3 instanceof h && t3.extend === it) {
		        if (this.fields && this.fields[t3.name] === t3) return delete this.fields[t3.name], t3.parent = null, t3.onRemove(this), r2(this);
		        throw Error(t3 + " is not a member of " + this);
		      }
		      if (t3 instanceof f) {
		        if (this.oneofs && this.oneofs[t3.name] === t3) return delete this.oneofs[t3.name], t3.parent = null, t3.onRemove(this), r2(this);
		        throw Error(t3 + " is not a member of " + this);
		      }
		      return o.prototype.remove.call(this, t3);
		    }, m.prototype.isReservedId = function(t3) {
		      return o.isReservedId(this.reserved, t3);
		    }, m.prototype.isReservedName = function(t3) {
		      return o.isReservedName(this.reserved, t3);
		    }, m.prototype.create = function(t3) {
		      return new this.ctor(t3);
		    }, m.prototype.setup = function() {
		      for (var t3 = this.fullName, i3 = [], n3 = 0; n3 < this.fieldsArray.length; ++n3) i3.push(this.e[n3].resolve().resolvedType);
		      this.encode = p(this)({ Writer: l, types: i3, util: d }), this.decode = v(this)({ Reader: s, types: i3, util: d }), this.verify = b(this)({ types: i3, util: d }), this.fromObject = w.fromObject(this)({ types: i3, util: d }), this.toObject = w.toObject(this)({ types: i3, util: d });
		      var r3, t3 = y[t3];
		      return t3 && ((r3 = Object.create(this)).fromObject = this.fromObject, this.fromObject = t3.fromObject.bind(r3), r3.toObject = this.toObject, this.toObject = t3.toObject.bind(r3)), this;
		    }, m.prototype.encode = function(t3, i3) {
		      return this.setup().encode(t3, i3);
		    }, m.prototype.encodeDelimited = function(t3, i3) {
		      return this.encode(t3, i3 && i3.len ? i3.fork() : i3).ldelim();
		    }, m.prototype.decode = function(t3, i3) {
		      return this.setup().decode(t3, i3);
		    }, m.prototype.decodeDelimited = function(t3) {
		      return t3 instanceof s || (t3 = s.create(t3)), this.decode(t3, t3.uint32());
		    }, m.prototype.verify = function(t3) {
		      return this.setup().verify(t3);
		    }, m.prototype.fromObject = function(t3) {
		      return this.setup().fromObject(t3);
		    }, m.prototype.toObject = function(t3, i3) {
		      return this.setup().toObject(t3, i3);
		    }, m.d = function(i3) {
		      return function(t3) {
		        d.decorateType(t3, i3);
		      };
		    };
		  }, { 12: 12, 13: 13, 14: 14, 15: 15, 16: 16, 20: 20, 21: 21, 23: 23, 25: 25, 27: 27, 33: 33, 37: 37, 40: 40, 41: 41, 42: 42 }], 36: [function(t2, i2, n2) {
		    var t2 = t2(37), e = ["double", "float", "int32", "uint32", "sint32", "fixed32", "sfixed32", "int64", "uint64", "sint64", "fixed64", "sfixed64", "bool", "string", "bytes"];
		    function r2(t3, i3) {
		      var n3 = 0, r3 = {};
		      for (i3 |= 0; n3 < t3.length; ) r3[e[n3 + i3]] = t3[n3++];
		      return r3;
		    }
		    n2.basic = r2([1, 5, 0, 0, 0, 5, 5, 0, 0, 0, 1, 1, 0, 2, 2]), n2.defaults = r2([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, false, "", t2.emptyArray, null]), n2.long = r2([0, 0, 0, 1, 1], 7), n2.mapKey = r2([0, 0, 0, 5, 5, 0, 0, 0, 1, 1, 0, 2], 2), n2.packed = r2([1, 5, 0, 0, 0, 5, 5, 0, 0, 0, 1, 1, 0]);
		  }, { 37: 37 }], 37: [function(n2, t2, i2) {
		    var r2, e, s = t2.exports = n2(39), o = n2(30), u = (s.codegen = n2(3), s.fetch = n2(5), s.path = n2(8), s.fs = s.inquire("fs"), s.toArray = function(t3) {
		      if (t3) {
		        for (var i3 = Object.keys(t3), n3 = Array(i3.length), r3 = 0; r3 < i3.length; ) n3[r3] = t3[i3[r3++]];
		        return n3;
		      }
		      return [];
		    }, s.toObject = function(t3) {
		      for (var i3 = {}, n3 = 0; n3 < t3.length; ) {
		        var r3 = t3[n3++], e2 = t3[n3++];
		        e2 !== it && (i3[r3] = e2);
		      }
		      return i3;
		    }, /\\/g), f = /"/g, h = (s.isReserved = function(t3) {
		      return /^(?:do|if|in|for|let|new|try|var|case|else|enum|eval|false|null|this|true|void|with|break|catch|class|const|super|throw|while|yield|delete|export|import|public|return|static|switch|typeof|default|extends|finally|package|private|continue|debugger|function|arguments|interface|protected|implements|instanceof)$/.test(t3);
		    }, s.safeProp = function(t3) {
		      return !/^[$\w_]+$/.test(t3) || s.isReserved(t3) ? '["' + t3.replace(u, "\\\\").replace(f, '\\"') + '"]' : "." + t3;
		    }, s.ucFirst = function(t3) {
		      return (t3[0] || "").toUpperCase() + t3.substring(1);
		    }, /_([a-z])/g), a = (s.camelCase = function(t3) {
		      return t3.substring(0, 1) + t3.substring(1).replace(h, function(t4, i3) {
		        return i3.toUpperCase();
		      });
		    }, s.compareFieldsById = function(t3, i3) {
		      return t3.id - i3.id;
		    }, s.decorateType = function(t3, i3) {
		      return t3.$type ? (i3 && t3.$type.name !== i3 && (s.decorateRoot.remove(t3.$type), t3.$type.name = i3, s.decorateRoot.add(t3.$type)), t3.$type) : (i3 = new (r2 = r2 || n2(35))(i3 || t3.name), s.decorateRoot.add(i3), i3.ctor = t3, Object.defineProperty(t3, "$type", { value: i3, enumerable: false }), Object.defineProperty(t3.prototype, "$type", { value: i3, enumerable: false }), i3);
		    }, 0);
		    s.decorateEnum = function(t3) {
		      var i3;
		      return t3.$type || (i3 = new (e = e || n2(15))("Enum" + a++, t3), s.decorateRoot.add(i3), Object.defineProperty(t3, "$type", { value: i3, enumerable: false }), i3);
		    }, s.setProperty = function(t3, i3, n3, s2) {
		      if ("object" != typeof t3) throw TypeError("dst must be an object");
		      if (i3) return (function t4(i4, n4, r3) {
		        var e2 = n4.shift();
		        if ("__proto__" !== e2 && "prototype" !== e2) if (0 < n4.length) i4[e2] = t4(i4[e2] || {}, n4, r3);
		        else {
		          if ((n4 = i4[e2]) && s2) return i4;
		          n4 && (r3 = [].concat(n4).concat(r3)), i4[e2] = r3;
		        }
		        return i4;
		      })(t3, i3 = i3.split("."), n3);
		      throw TypeError("path must be specified");
		    }, Object.defineProperty(s, "decorateRoot", { get: function() {
		      return o.decorated || (o.decorated = new (n2(29))());
		    } });
		  }, { 15: 15, 29: 29, 3: 3, 30: 30, 35: 35, 39: 39, 5: 5, 8: 8 }], 38: [function(t2, i2, n2) {
		    i2.exports = e;
		    var r2 = t2(39);
		    function e(t3, i3) {
		      this.lo = t3 >>> 0, this.hi = i3 >>> 0;
		    }
		    var s = e.zero = new e(0, 0), o = (s.toNumber = function() {
		      return 0;
		    }, s.zzEncode = s.zzDecode = function() {
		      return this;
		    }, s.length = function() {
		      return 1;
		    }, e.zeroHash = "\0\0\0\0\0\0\0\0", e.fromNumber = function(t3) {
		      var i3, n3;
		      return 0 === t3 ? s : (n3 = (t3 = (i3 = t3 < 0) ? -t3 : t3) >>> 0, t3 = (t3 - n3) / 4294967296 >>> 0, i3 && (t3 = ~t3 >>> 0, n3 = ~n3 >>> 0, 4294967295 < ++n3 && (n3 = 0, 4294967295 < ++t3 && (t3 = 0))), new e(n3, t3));
		    }, e.from = function(t3) {
		      if ("number" == typeof t3) return e.fromNumber(t3);
		      if (r2.isString(t3)) {
		        if (!r2.Long) return e.fromNumber(parseInt(t3, 10));
		        t3 = r2.Long.fromString(t3);
		      }
		      return t3.low || t3.high ? new e(t3.low >>> 0, t3.high >>> 0) : s;
		    }, e.prototype.toNumber = function(t3) {
		      var i3;
		      return !t3 && this.hi >>> 31 ? (t3 = 1 + ~this.lo >>> 0, i3 = ~this.hi >>> 0, -(t3 + 4294967296 * (i3 = t3 ? i3 : i3 + 1 >>> 0))) : this.lo + 4294967296 * this.hi;
		    }, e.prototype.toLong = function(t3) {
		      return r2.Long ? new r2.Long(0 | this.lo, 0 | this.hi, !!t3) : { low: 0 | this.lo, high: 0 | this.hi, unsigned: !!t3 };
		    }, String.prototype.charCodeAt);
		    e.fromHash = function(t3) {
		      return "\0\0\0\0\0\0\0\0" === t3 ? s : new e((o.call(t3, 0) | o.call(t3, 1) << 8 | o.call(t3, 2) << 16 | o.call(t3, 3) << 24) >>> 0, (o.call(t3, 4) | o.call(t3, 5) << 8 | o.call(t3, 6) << 16 | o.call(t3, 7) << 24) >>> 0);
		    }, e.prototype.toHash = function() {
		      return String.fromCharCode(255 & this.lo, this.lo >>> 8 & 255, this.lo >>> 16 & 255, this.lo >>> 24, 255 & this.hi, this.hi >>> 8 & 255, this.hi >>> 16 & 255, this.hi >>> 24);
		    }, e.prototype.zzEncode = function() {
		      var t3 = this.hi >> 31;
		      return this.hi = ((this.hi << 1 | this.lo >>> 31) ^ t3) >>> 0, this.lo = (this.lo << 1 ^ t3) >>> 0, this;
		    }, e.prototype.zzDecode = function() {
		      var t3 = -(1 & this.lo);
		      return this.lo = ((this.lo >>> 1 | this.hi << 31) ^ t3) >>> 0, this.hi = (this.hi >>> 1 ^ t3) >>> 0, this;
		    }, e.prototype.length = function() {
		      var t3 = this.lo, i3 = (this.lo >>> 28 | this.hi << 4) >>> 0, n3 = this.hi >>> 24;
		      return 0 == n3 ? 0 == i3 ? t3 < 16384 ? t3 < 128 ? 1 : 2 : t3 < 2097152 ? 3 : 4 : i3 < 16384 ? i3 < 128 ? 5 : 6 : i3 < 2097152 ? 7 : 8 : n3 < 128 ? 9 : 10;
		    };
		  }, { 39: 39 }], 39: [function(t2, i2, n2) {
		    var r2 = n2;
		    function e(t3, i3, n3) {
		      for (var r3 = Object.keys(i3), e2 = 0; e2 < r3.length; ++e2) t3[r3[e2]] !== it && n3 || (t3[r3[e2]] = i3[r3[e2]]);
		      return t3;
		    }
		    function s(t3) {
		      function n3(t4, i3) {
		        if (!(this instanceof n3)) return new n3(t4, i3);
		        Object.defineProperty(this, "message", { get: function() {
		          return t4;
		        } }), Error.captureStackTrace ? Error.captureStackTrace(this, n3) : Object.defineProperty(this, "stack", { value: Error().stack || "" }), i3 && e(this, i3);
		      }
		      return n3.prototype = Object.create(Error.prototype, { constructor: { value: n3, writable: true, enumerable: false, configurable: true }, name: { get: function() {
		        return t3;
		      }, set: it, enumerable: false, configurable: true }, toString: { value: function() {
		        return this.name + ": " + this.message;
		      }, writable: true, enumerable: false, configurable: true } }), n3;
		    }
		    r2.asPromise = t2(1), r2.base64 = t2(2), r2.EventEmitter = t2(4), r2.float = t2(6), r2.inquire = t2(7), r2.utf8 = t2(10), r2.pool = t2(9), r2.LongBits = t2(38), r2.isNode = !!("undefined" != typeof globalThis && globalThis && globalThis.process && globalThis.process.versions && globalThis.process.versions.node), r2.global = r2.isNode && globalThis || "undefined" != typeof window && window || "undefined" != typeof self && self || this, r2.emptyArray = Object.freeze ? Object.freeze([]) : [], r2.emptyObject = Object.freeze ? Object.freeze({}) : {}, r2.isInteger = Number.isInteger || function(t3) {
		      return "number" == typeof t3 && isFinite(t3) && Math.floor(t3) === t3;
		    }, r2.isString = function(t3) {
		      return "string" == typeof t3 || t3 instanceof String;
		    }, r2.isObject = function(t3) {
		      return t3 && "object" == typeof t3;
		    }, r2.isset = r2.isSet = function(t3, i3) {
		      var n3 = t3[i3];
		      return null != n3 && t3.hasOwnProperty(i3) && ("object" != typeof n3 || 0 < (Array.isArray(n3) ? n3 : Object.keys(n3)).length);
		    }, r2.Buffer = (function() {
		      try {
		        var t3 = r2.inquire("buffer").Buffer;
		        return t3.prototype.utf8Write ? t3 : null;
		      } catch (t4) {
		        return null;
		      }
		    })(), r2.N = null, r2.P = null, r2.newBuffer = function(t3) {
		      return "number" == typeof t3 ? r2.Buffer ? r2.P(t3) : new r2.Array(t3) : r2.Buffer ? r2.N(t3) : "undefined" == typeof Uint8Array ? t3 : new Uint8Array(t3);
		    }, r2.Array = "undefined" != typeof Uint8Array ? Uint8Array : Array, r2.Long = r2.global.dcodeIO && r2.global.dcodeIO.Long || r2.global.Long || r2.inquire("long"), r2.key2Re = /^true|false|0|1$/, r2.key32Re = /^-?(?:0|[1-9][0-9]*)$/, r2.key64Re = /^(?:[\\x00-\\xff]{8}|-?(?:0|[1-9][0-9]*))$/, r2.longToHash = function(t3) {
		      return t3 ? r2.LongBits.from(t3).toHash() : r2.LongBits.zeroHash;
		    }, r2.longFromHash = function(t3, i3) {
		      t3 = r2.LongBits.fromHash(t3);
		      return r2.Long ? r2.Long.fromBits(t3.lo, t3.hi, i3) : t3.toNumber(!!i3);
		    }, r2.merge = e, r2.lcFirst = function(t3) {
		      return (t3[0] || "").toLowerCase() + t3.substring(1);
		    }, r2.newError = s, r2.ProtocolError = s("ProtocolError"), r2.oneOfGetter = function(t3) {
		      for (var n3 = {}, i3 = 0; i3 < t3.length; ++i3) n3[t3[i3]] = 1;
		      return function() {
		        for (var t4 = Object.keys(this), i4 = t4.length - 1; -1 < i4; --i4) if (1 === n3[t4[i4]] && this[t4[i4]] !== it && null !== this[t4[i4]]) return t4[i4];
		      };
		    }, r2.oneOfSetter = function(n3) {
		      return function(t3) {
		        for (var i3 = 0; i3 < n3.length; ++i3) n3[i3] !== t3 && delete this[n3[i3]];
		      };
		    }, r2.toJSONOptions = { longs: String, enums: String, bytes: String, json: true }, r2.p = function() {
		      var n3 = r2.Buffer;
		      n3 ? (r2.N = n3.from !== Uint8Array.from && n3.from || function(t3, i3) {
		        return new n3(t3, i3);
		      }, r2.P = n3.allocUnsafe || function(t3) {
		        return new n3(t3);
		      }) : r2.N = r2.P = null;
		    };
		  }, { 1: 1, 10: 10, 2: 2, 38: 38, 4: 4, 6: 6, 7: 7, 9: 9 }], 40: [function(t2, i2, n2) {
		    i2.exports = function(t3) {
		      var i3 = f.codegen(["m"], t3.name + "$verify")('if(typeof m!=="object"||m===null)')("return%j", "object expected"), n3 = t3.oneofsArray, r2 = {};
		      n3.length && i3("var p={}");
		      for (var e = 0; e < t3.fieldsArray.length; ++e) {
		        var s, o2 = t3.e[e].resolve(), u = "m" + f.safeProp(o2.name);
		        o2.optional && i3("if(%s!=null&&m.hasOwnProperty(%j)){", u, o2.name), o2.map ? (i3("if(!util.isObject(%s))", u)("return%j", h(o2, "object"))("var k=Object.keys(%s)", u)("for(var i=0;i<k.length;++i){"), (function(t4, i4, n4) {
		          switch (i4.keyType) {
		            case "int32":
		            case "uint32":
		            case "sint32":
		            case "fixed32":
		            case "sfixed32":
		              t4("if(!util.key32Re.test(%s))", n4)("return%j", h(i4, "integer key"));
		              break;
		            case "int64":
		            case "uint64":
		            case "sint64":
		            case "fixed64":
		            case "sfixed64":
		              t4("if(!util.key64Re.test(%s))", n4)("return%j", h(i4, "integer|Long key"));
		              break;
		            case "bool":
		              t4("if(!util.key2Re.test(%s))", n4)("return%j", h(i4, "boolean key"));
		          }
		        })(i3, o2, "k[i]"), a(i3, o2, e, u + "[k[i]]")("}")) : o2.repeated ? (i3("if(!Array.isArray(%s))", u)("return%j", h(o2, "array"))("for(var i=0;i<%s.length;++i){", u), a(i3, o2, e, u + "[i]")("}")) : (o2.partOf && (s = f.safeProp(o2.partOf.name), 1 === r2[o2.partOf.name] && i3("if(p%s===1)", s)("return%j", o2.partOf.name + ": multiple values"), r2[o2.partOf.name] = 1, i3("p%s=1", s)), a(i3, o2, e, u)), o2.optional && i3("}");
		      }
		      return i3("return null");
		    };
		    var o = t2(15), f = t2(37);
		    function h(t3, i3) {
		      return t3.name + ": " + i3 + (t3.repeated && "array" !== i3 ? "[]" : t3.map && "object" !== i3 ? "{k:" + t3.keyType + "}" : "") + " expected";
		    }
		    function a(t3, i3, n3, r2) {
		      if (i3.resolvedType) if (i3.resolvedType instanceof o) {
		        t3("switch(%s){", r2)("default:")("return%j", h(i3, "enum value"));
		        for (var e = Object.keys(i3.resolvedType.values), s = 0; s < e.length; ++s) t3("case %i:", i3.resolvedType.values[e[s]]);
		        t3("break")("}");
		      } else t3("{")("var e=types[%i].verify(%s);", n3, r2)("if(e)")("return%j+e", i3.name + ".")("}");
		      else switch (i3.type) {
		        case "int32":
		        case "uint32":
		        case "sint32":
		        case "fixed32":
		        case "sfixed32":
		          t3("if(!util.isInteger(%s))", r2)("return%j", h(i3, "integer"));
		          break;
		        case "int64":
		        case "uint64":
		        case "sint64":
		        case "fixed64":
		        case "sfixed64":
		          t3("if(!util.isInteger(%s)&&!(%s&&util.isInteger(%s.low)&&util.isInteger(%s.high)))", r2, r2, r2, r2)("return%j", h(i3, "integer|Long"));
		          break;
		        case "float":
		        case "double":
		          t3('if(typeof %s!=="number")', r2)("return%j", h(i3, "number"));
		          break;
		        case "bool":
		          t3('if(typeof %s!=="boolean")', r2)("return%j", h(i3, "boolean"));
		          break;
		        case "string":
		          t3("if(!util.isString(%s))", r2)("return%j", h(i3, "string"));
		          break;
		        case "bytes":
		          t3('if(!(%s&&typeof %s.length==="number"||util.isString(%s)))', r2, r2, r2)("return%j", h(i3, "buffer"));
		      }
		      return t3;
		    }
		  }, { 15: 15, 37: 37 }], 41: [function(t2, i2, n2) {
		    var o = t2(21);
		    n2[".google.protobuf.Any"] = { fromObject: function(t3) {
		      if (t3 && t3["@type"]) {
		        var i3, n3 = t3["@type"].substring(1 + t3["@type"].lastIndexOf("/")), n3 = this.lookup(n3);
		        if (n3) return ~(i3 = "." == (t3["@type"][0] || "") ? t3["@type"].slice(1) : t3["@type"]).indexOf("/") || (i3 = "/" + i3), this.create({ type_url: i3, value: n3.encode(n3.fromObject(t3)).finish() });
		      }
		      return this.fromObject(t3);
		    }, toObject: function(t3, i3) {
		      var n3, r2, e = "", s = "";
		      return i3 && i3.json && t3.type_url && t3.value && (s = t3.type_url.substring(1 + t3.type_url.lastIndexOf("/")), e = t3.type_url.substring(0, 1 + t3.type_url.lastIndexOf("/")), (n3 = this.lookup(s)) && (t3 = n3.decode(t3.value))), !(t3 instanceof this.ctor) && t3 instanceof o ? (n3 = t3.$type.toObject(t3, i3), r2 = "." === t3.$type.fullName[0] ? t3.$type.fullName.slice(1) : t3.$type.fullName, n3["@type"] = s = (e = "" === e ? "type.googleapis.com/" : e) + r2, n3) : this.toObject(t3, i3);
		    } };
		  }, { 21: 21 }], 42: [function(t2, i2, n2) {
		    i2.exports = c;
		    var r2, e = t2(39), s = e.LongBits, o = e.base64, u = e.utf8;
		    function f(t3, i3, n3) {
		      this.fn = t3, this.len = i3, this.next = it, this.val = n3;
		    }
		    function h() {
		    }
		    function a(t3) {
		      this.head = t3.head, this.tail = t3.tail, this.len = t3.len, this.next = t3.states;
		    }
		    function c() {
		      this.len = 0, this.head = new f(h, 0, 0), this.tail = this.head, this.states = null;
		    }
		    function l() {
		      return e.Buffer ? function() {
		        return (c.create = function() {
		          return new r2();
		        })();
		      } : function() {
		        return new c();
		      };
		    }
		    function d(t3, i3, n3) {
		      i3[n3] = 255 & t3;
		    }
		    function p(t3, i3) {
		      this.len = t3, this.next = it, this.val = i3;
		    }
		    function v(t3, i3, n3) {
		      for (; t3.hi; ) i3[n3++] = 127 & t3.lo | 128, t3.lo = (t3.lo >>> 7 | t3.hi << 25) >>> 0, t3.hi >>>= 7;
		      for (; 127 < t3.lo; ) i3[n3++] = 127 & t3.lo | 128, t3.lo = t3.lo >>> 7;
		      i3[n3++] = t3.lo;
		    }
		    function b(t3, i3, n3) {
		      i3[n3] = 255 & t3, i3[n3 + 1] = t3 >>> 8 & 255, i3[n3 + 2] = t3 >>> 16 & 255, i3[n3 + 3] = t3 >>> 24;
		    }
		    c.create = l(), c.alloc = function(t3) {
		      return new e.Array(t3);
		    }, e.Array !== Array && (c.alloc = e.pool(c.alloc, e.Array.prototype.subarray)), c.prototype.V = function(t3, i3, n3) {
		      return this.tail = this.tail.next = new f(t3, i3, n3), this.len += i3, this;
		    }, (p.prototype = Object.create(f.prototype)).fn = function(t3, i3, n3) {
		      for (; 127 < t3; ) i3[n3++] = 127 & t3 | 128, t3 >>>= 7;
		      i3[n3] = t3;
		    }, c.prototype.uint32 = function(t3) {
		      return this.len += (this.tail = this.tail.next = new p((t3 >>>= 0) < 128 ? 1 : t3 < 16384 ? 2 : t3 < 2097152 ? 3 : t3 < 268435456 ? 4 : 5, t3)).len, this;
		    }, c.prototype.int32 = function(t3) {
		      return t3 < 0 ? this.V(v, 10, s.fromNumber(t3)) : this.uint32(t3);
		    }, c.prototype.sint32 = function(t3) {
		      return this.uint32((t3 << 1 ^ t3 >> 31) >>> 0);
		    }, c.prototype.int64 = c.prototype.uint64 = function(t3) {
		      t3 = s.from(t3);
		      return this.V(v, t3.length(), t3);
		    }, c.prototype.sint64 = function(t3) {
		      t3 = s.from(t3).zzEncode();
		      return this.V(v, t3.length(), t3);
		    }, c.prototype.bool = function(t3) {
		      return this.V(d, 1, t3 ? 1 : 0);
		    }, c.prototype.sfixed32 = c.prototype.fixed32 = function(t3) {
		      return this.V(b, 4, t3 >>> 0);
		    }, c.prototype.sfixed64 = c.prototype.fixed64 = function(t3) {
		      t3 = s.from(t3);
		      return this.V(b, 4, t3.lo).V(b, 4, t3.hi);
		    }, c.prototype.float = function(t3) {
		      return this.V(e.float.writeFloatLE, 4, t3);
		    }, c.prototype.double = function(t3) {
		      return this.V(e.float.writeDoubleLE, 8, t3);
		    };
		    var w = e.Array.prototype.set ? function(t3, i3, n3) {
		      i3.set(t3, n3);
		    } : function(t3, i3, n3) {
		      for (var r3 = 0; r3 < t3.length; ++r3) i3[n3 + r3] = t3[r3];
		    };
		    c.prototype.bytes = function(t3) {
		      var i3, n3 = t3.length >>> 0;
		      return n3 ? (e.isString(t3) && (i3 = c.alloc(n3 = o.length(t3)), o.decode(t3, i3, 0), t3 = i3), this.uint32(n3).V(w, n3, t3)) : this.V(d, 1, 0);
		    }, c.prototype.string = function(t3) {
		      var i3 = u.length(t3);
		      return i3 ? this.uint32(i3).V(u.write, i3, t3) : this.V(d, 1, 0);
		    }, c.prototype.fork = function() {
		      return this.states = new a(this), this.head = this.tail = new f(h, 0, 0), this.len = 0, this;
		    }, c.prototype.reset = function() {
		      return this.states ? (this.head = this.states.head, this.tail = this.states.tail, this.len = this.states.len, this.states = this.states.next) : (this.head = this.tail = new f(h, 0, 0), this.len = 0), this;
		    }, c.prototype.ldelim = function() {
		      var t3 = this.head, i3 = this.tail, n3 = this.len;
		      return this.reset().uint32(n3), n3 && (this.tail.next = t3.next, this.tail = i3, this.len += n3), this;
		    }, c.prototype.finish = function() {
		      for (var t3 = this.head.next, i3 = this.constructor.alloc(this.len), n3 = 0; t3; ) t3.fn(t3.val, i3, n3), n3 += t3.len, t3 = t3.next;
		      return i3;
		    }, c.p = function(t3) {
		      r2 = t3, c.create = l(), r2.p();
		    };
		  }, { 39: 39 }], 43: [function(t2, i2, n2) {
		    i2.exports = s;
		    var r2 = t2(42), e = ((s.prototype = Object.create(r2.prototype)).constructor = s, t2(39));
		    function s() {
		      r2.call(this);
		    }
		    function o(t3, i3, n3) {
		      t3.length < 40 ? e.utf8.write(t3, i3, n3) : i3.utf8Write ? i3.utf8Write(t3, n3) : i3.write(t3, n3);
		    }
		    s.p = function() {
		      s.alloc = e.P, s.writeBytesBuffer = e.Buffer && e.Buffer.prototype instanceof Uint8Array && "set" === e.Buffer.prototype.set.name ? function(t3, i3, n3) {
		        i3.set(t3, n3);
		      } : function(t3, i3, n3) {
		        if (t3.copy) t3.copy(i3, n3, 0, t3.length);
		        else for (var r3 = 0; r3 < t3.length; ) i3[n3++] = t3[r3++];
		      };
		    }, s.prototype.bytes = function(t3) {
		      var i3 = (t3 = e.isString(t3) ? e.N(t3, "base64") : t3).length >>> 0;
		      return this.uint32(i3), i3 && this.V(s.writeBytesBuffer, i3, t3), this;
		    }, s.prototype.string = function(t3) {
		      var i3 = e.Buffer.byteLength(t3);
		      return this.uint32(i3), i3 && this.V(o, i3, t3), this;
		    }, s.p();
		  }, { 39: 39, 42: 42 }] }, {}, [19]);
		})(); 
	} (protobuf_min));
	return protobuf_min.exports;
}

var protobuf_minExports = requireProtobuf_min();
const protobuf = /*@__PURE__*/getDefaultExportFromCjs(protobuf_minExports);

export { protobuf as p };
