// @btc-vision/bitcoin/build/payments/embed.js
/**
 * OP_RETURN Embed payment class.
 *
 * Embed payments use OP_RETURN to store arbitrary data in the blockchain.
 * These outputs are provably unspendable.
 *
 * @packageDocumentation
 */
import { bitcoin as BITCOIN_NETWORK } from '../networks.js';
import * as bscript from '../script.js';
import { stacksEqual } from '../types.js';
import { PaymentType } from './types.js';
const OPS = bscript.opcodes;
/**
 * OP_RETURN Embed payment class.
 *
 * Creates outputs of the form: `OP_RETURN {data1} {data2} ...`
 * These outputs are provably unspendable and used for data storage.
 *
 * @example
 * ```typescript
 * import { Embed } from '@btc-vision/bitcoin';
 *
 * // Create from data
 * const payment = Embed.fromData([Buffer.from('Hello, Bitcoin!')]);
 * console.log(payment.output); // scriptPubKey with OP_RETURN
 *
 * // Decode an existing output
 * const decoded = Embed.fromOutput(scriptPubKey);
 * console.log(decoded.data); // array of data chunks
 * ```
 */
export class Embed {
    // Static public fields
    static NAME = PaymentType.Embed;
    // Private instance fields
    #network;
    #opts;
    // Input data (provided by user)
    #inputData;
    #inputOutput;
    // Cached computed values
    #data;
    #output;
    // Cache flags
    #dataComputed = false;
    #outputComputed = false;
    /**
     * Creates a new Embed payment instance.
     *
     * @param params - Payment parameters
     * @param params.data - Array of data chunks to embed
     * @param params.output - The scriptPubKey
     * @param params.network - Network parameters (defaults to mainnet)
     * @param opts - Payment options
     * @param opts.validate - Whether to validate inputs (default: true)
     *
     * @throws {TypeError} If validation is enabled and data is invalid
     */
    constructor(params, opts) {
        this.#network = params.network ?? BITCOIN_NETWORK;
        this.#opts = {
            validate: opts?.validate ?? true,
            allowIncomplete: opts?.allowIncomplete ?? false,
        };
        // Store input data
        this.#inputData = params.data;
        this.#inputOutput = params.output;
        // Validate if requested
        if (this.#opts.validate) {
            this.#validate();
        }
    }
    // Public getters
    /**
     * Payment type discriminant.
     */
    get name() {
        return PaymentType.Embed;
    }
    /**
     * Network parameters.
     */
    get network() {
        return this.#network;
    }
    /**
     * The embedded data chunks.
     */
    get data() {
        if (!this.#dataComputed) {
            this.#data = this.#computeData();
            this.#dataComputed = true;
        }
        return this.#data ?? [];
    }
    /**
     * The scriptPubKey: `OP_RETURN {data...}`
     */
    get output() {
        if (!this.#outputComputed) {
            this.#output = this.#computeOutput();
            this.#outputComputed = true;
        }
        return this.#output;
    }
    // Static factory methods
    /**
     * Creates an Embed payment from data chunks.
     *
     * @param data - Array of data chunks to embed
     * @param network - Network parameters (defaults to mainnet)
     * @returns A new Embed payment instance
     *
     * @example
     * ```typescript
     * const payment = Embed.fromData([
     *   new TextEncoder().encode('Hello'),
     *   new TextEncoder().encode('Bitcoin')
     * ]);
     * ```
     */
    static fromData(data, network) {
        return new Embed({ data, network });
    }
    /**
     * Creates an Embed payment from a scriptPubKey.
     *
     * @param output - The scriptPubKey
     * @param network - Network parameters (defaults to mainnet)
     * @returns A new Embed payment instance
     */
    static fromOutput(output, network) {
        return new Embed({ output, network });
    }
    // Private computation methods
    /**
     * Converts to a plain EmbedPayment object for backwards compatibility.
     *
     * @returns An EmbedPayment object
     */
    toPayment() {
        return {
            name: this.name,
            network: this.network,
            data: this.data,
            output: this.output,
        };
    }
    #computeData() {
        if (this.#inputData) {
            return this.#inputData;
        }
        if (this.#inputOutput) {
            const script = bscript.decompile(this.#inputOutput);
            if (script === null || script === undefined) {
                return undefined;
            }
            return script.slice(1);
        }
        return undefined;
    }
    // Validation
    #computeOutput() {
        if (this.#inputOutput) {
            return this.#inputOutput;
        }
        if (!this.#inputData) {
            return undefined;
        }
        return bscript.compile([OPS.OP_RETURN].concat(this.#inputData));
    }
    #validate() {
        if (this.#inputOutput) {
            const chunks = bscript.decompile(this.#inputOutput);
            if (!chunks) {
                throw new TypeError('Output is invalid');
            }
            if (chunks[0] !== OPS.OP_RETURN) {
                throw new TypeError('Output is invalid');
            }
            if (!chunks.slice(1).every((c) => c instanceof Uint8Array)) {
                throw new TypeError('Output is invalid');
            }
            if (this.#inputData && !stacksEqual(this.#inputData, this.data)) {
                throw new TypeError('Data mismatch');
            }
        }
    }
}
/**
 * Creates an OP_RETURN Embed payment object.
 *
 * This is the legacy factory function for backwards compatibility.
 * For new code, prefer using the Embed class directly.
 *
 * @param a - The payment object containing the necessary data
 * @param opts - Optional payment options
 * @returns The Embed payment object
 * @throws {TypeError} If the required data is not provided or if the data is invalid
 *
 * @example
 * ```typescript
 * import { p2data } from '@btc-vision/bitcoin';
 *
 * // Create from data
 * const payment = p2data({ data: [Buffer.from('Hello')] });
 *
 * // Decode from output
 * const decoded = p2data({ output: scriptPubKey });
 * ```
 */
export function p2data(a, opts) {
    if (!a.data && !a.output) {
        throw new TypeError('Not enough data');
    }
    const instance = new Embed({
        data: a.data,
        output: a.output,
        network: a.network,
    }, opts);
    // Return a merged object for backwards compatibility
    return Object.assign(instance.toPayment(), a);
}
