// @btc-vision/bitcoin/build/payments/p2pk.js
/**
 * Pay-to-Public-Key (P2PK) payment class.
 *
 * P2PK is one of the simplest Bitcoin payment types where the output script
 * contains a public key directly, and spending requires a signature from that key.
 *
 * @packageDocumentation
 */
import { bitcoin as BITCOIN_NETWORK } from '../networks.js';
import * as bscript from '../script.js';
import { isPoint } from '../types.js';
import { equals } from '../io/index.js';
import { PaymentType } from './types.js';
const OPS = bscript.opcodes;
/**
 * Pay-to-Public-Key (P2PK) payment class.
 *
 * Creates locking scripts of the form: `{pubKey} OP_CHECKSIG`
 * Spending requires providing: `{signature}`
 *
 * @example
 * ```typescript
 * import { P2PK } from '@btc-vision/bitcoin';
 *
 * // Create from public key
 * const payment = P2PK.fromPubkey(pubkey);
 * console.log(payment.output); // scriptPubKey
 *
 * // Create from output script
 * const decoded = P2PK.fromOutput(scriptPubKey);
 * console.log(decoded.pubkey); // extracted public key
 *
 * // Legacy factory function still works
 * const legacy = p2pk({ pubkey });
 * ```
 */
export class P2PK {
    // Static public fields
    static NAME = PaymentType.P2PK;
    // Private instance fields
    #network;
    #opts;
    // Input data (provided by user)
    #inputPubkey;
    #inputSignature;
    #inputOutput;
    #inputInput;
    // Cached computed values
    #pubkey;
    #signature;
    #output;
    #input;
    #witness;
    // Cache flags
    #pubkeyComputed = false;
    #signatureComputed = false;
    #outputComputed = false;
    #inputComputed = false;
    #witnessComputed = false;
    /**
     * Creates a new P2PK payment instance.
     *
     * @param params - Payment parameters
     * @param params.pubkey - The public key (33 or 65 bytes)
     * @param params.signature - DER-encoded signature
     * @param params.output - The scriptPubKey
     * @param params.input - The scriptSig
     * @param params.network - Network parameters (defaults to mainnet)
     * @param opts - Payment options
     * @param opts.validate - Whether to validate inputs (default: true)
     *
     * @throws {TypeError} If validation is enabled and data is invalid
     */
    constructor(params, opts) {
        this.#network = params.network ?? BITCOIN_NETWORK;
        this.#opts = {
            validate: opts?.validate ?? true,
            allowIncomplete: opts?.allowIncomplete ?? false,
        };
        // Store input data
        this.#inputPubkey = params.pubkey;
        this.#inputSignature = params.signature;
        this.#inputOutput = params.output;
        this.#inputInput = params.input;
        // Validate if requested
        if (this.#opts.validate) {
            this.#validate();
        }
    }
    // Public getters
    /**
     * Payment type discriminant.
     */
    get name() {
        return PaymentType.P2PK;
    }
    /**
     * Network parameters.
     */
    get network() {
        return this.#network;
    }
    /**
     * The public key (33 or 65 bytes).
     * Computed lazily from output if not provided directly.
     */
    get pubkey() {
        if (!this.#pubkeyComputed) {
            this.#pubkey = this.#computePubkey();
            this.#pubkeyComputed = true;
        }
        return this.#pubkey;
    }
    /**
     * The DER-encoded signature.
     * Computed lazily from input if not provided directly.
     */
    get signature() {
        if (!this.#signatureComputed) {
            this.#signature = this.#computeSignature();
            this.#signatureComputed = true;
        }
        return this.#signature;
    }
    /**
     * The scriptPubKey: `{pubKey} OP_CHECKSIG`
     * Computed lazily from pubkey if not provided directly.
     */
    get output() {
        if (!this.#outputComputed) {
            this.#output = this.#computeOutput();
            this.#outputComputed = true;
        }
        return this.#output;
    }
    /**
     * The scriptSig: `{signature}`
     * Computed lazily from signature if not provided directly.
     */
    get input() {
        if (!this.#inputComputed) {
            this.#input = this.#computeInput();
            this.#inputComputed = true;
        }
        return this.#input;
    }
    /**
     * Witness stack (empty for P2PK as it's not a SegWit type).
     */
    get witness() {
        if (!this.#witnessComputed) {
            this.#witness = this.#computeWitness();
            this.#witnessComputed = true;
        }
        return this.#witness;
    }
    // Static factory methods
    /**
     * Creates a P2PK payment from a public key.
     *
     * @param pubkey - The public key (33 or 65 bytes)
     * @param network - Network parameters (defaults to mainnet)
     * @returns A new P2PK payment instance
     *
     * @example
     * ```typescript
     * const payment = P2PK.fromPubkey(pubkey);
     * const scriptPubKey = payment.output;
     * ```
     */
    static fromPubkey(pubkey, network) {
        return new P2PK({ pubkey, network });
    }
    /**
     * Creates a P2PK payment from a scriptPubKey.
     *
     * @param output - The scriptPubKey
     * @param network - Network parameters (defaults to mainnet)
     * @returns A new P2PK payment instance
     *
     * @example
     * ```typescript
     * const payment = P2PK.fromOutput(scriptPubKey);
     * const pubkey = payment.pubkey;
     * ```
     */
    static fromOutput(output, network) {
        return new P2PK({ output, network });
    }
    /**
     * Creates a P2PK payment from a signature (for spending).
     *
     * @param signature - The DER-encoded signature
     * @param pubkey - The public key (optional, for validation)
     * @param network - Network parameters (defaults to mainnet)
     * @returns A new P2PK payment instance
     *
     * @example
     * ```typescript
     * const payment = P2PK.fromSignature(signature, pubkey);
     * const scriptSig = payment.input;
     * ```
     */
    static fromSignature(signature, pubkey, network) {
        return new P2PK({ signature, pubkey, network });
    }
    // Private computation methods
    /**
     * Converts to a plain P2PKPayment object for backwards compatibility.
     *
     * @returns A P2PKPayment object
     */
    toPayment() {
        return {
            name: this.name,
            network: this.network,
            pubkey: this.pubkey,
            signature: this.signature,
            output: this.output,
            input: this.input,
            witness: this.witness,
        };
    }
    #computePubkey() {
        if (this.#inputPubkey) {
            return this.#inputPubkey;
        }
        if (this.#inputOutput) {
            // Extract pubkey from output: {pubkey} OP_CHECKSIG
            return this.#inputOutput.subarray(1, -1);
        }
        return undefined;
    }
    #computeSignature() {
        if (this.#inputSignature) {
            return this.#inputSignature;
        }
        if (this.#inputInput) {
            const chunks = bscript.decompile(this.#inputInput);
            if (chunks && chunks.length > 0) {
                return chunks[0];
            }
        }
        return undefined;
    }
    #computeOutput() {
        if (this.#inputOutput) {
            return this.#inputOutput;
        }
        const pubkey = this.#inputPubkey;
        if (pubkey) {
            return bscript.compile([pubkey, OPS.OP_CHECKSIG]);
        }
        return undefined;
    }
    #computeInput() {
        if (this.#inputInput) {
            return this.#inputInput;
        }
        const signature = this.#inputSignature;
        if (signature) {
            return bscript.compile([signature]);
        }
        return undefined;
    }
    // Validation
    #computeWitness() {
        if (this.input) {
            return [];
        }
        return undefined;
    }
    #validate() {
        if (this.#inputOutput) {
            if (this.#inputOutput[this.#inputOutput.length - 1] !== OPS.OP_CHECKSIG) {
                throw new TypeError('Output is invalid');
            }
            const extractedPubkey = this.pubkey;
            if (!isPoint(extractedPubkey)) {
                throw new TypeError('Output pubkey is invalid');
            }
            if (this.#inputPubkey &&
                extractedPubkey &&
                !equals(this.#inputPubkey, extractedPubkey)) {
                throw new TypeError('Pubkey mismatch');
            }
        }
        if (this.#inputSignature) {
            const computedInput = this.input;
            if (this.#inputInput && computedInput && !equals(this.#inputInput, computedInput)) {
                throw new TypeError('Signature mismatch');
            }
        }
        if (this.#inputInput) {
            const chunks = bscript.decompile(this.#inputInput);
            if (!chunks || chunks.length !== 1) {
                throw new TypeError('Input is invalid');
            }
            const sig = this.signature;
            if (!sig || !bscript.isCanonicalScriptSignature(sig)) {
                throw new TypeError('Input has invalid signature');
            }
        }
    }
}
/**
 * Creates a Pay-to-Public-Key (P2PK) payment object.
 *
 * This is the legacy factory function for backwards compatibility.
 * For new code, prefer using the P2PK class directly.
 *
 * @param a - The payment object containing the necessary data
 * @param opts - Optional payment options
 * @returns The P2PK payment object
 * @throws {TypeError} If the required data is not provided or if the data is invalid
 *
 * @example
 * ```typescript
 * import { p2pk } from '@btc-vision/bitcoin';
 *
 * // Create from public key
 * const payment = p2pk({ pubkey });
 *
 * // Create from output
 * const decoded = p2pk({ output: scriptPubKey });
 * ```
 */
export function p2pk(a, opts) {
    if (!a.input && !a.output && !a.pubkey && !a.signature) {
        throw new TypeError('Not enough data');
    }
    const instance = new P2PK({
        pubkey: a.pubkey,
        signature: a.signature,
        output: a.output,
        input: a.input,
        network: a.network,
    }, opts);
    // Return a merged object for backwards compatibility
    return Object.assign(instance.toPayment(), a);
}
